{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create TCGA Failed Sample Data Sets\n",
    "Several submissions fail. Typically this is because Bam2Fastq ran out of memory. Use the Sample.tsv to find the failed sample and create a new sample set so that we can re-run with out accidently causing needless recompute of succesful samples. \n",
    "\n",
    "change wdl cause everything to recompujte\n",
    "\n",
    "- ref:\n",
    "- [Make a set of data  table from scratch ](https://support.terra.bio/hc/en-us/articles/360047611871#h_01EJXZMM6GA3481YRRQBR65MY3)\n",
    "\n",
    "- [Data Tables QuickStart Part 3: Understanding sets of data](https://support.terra.bio/hc/en-us/articles/360047611871)\n",
    "\n",
    "- [adding data to a workspace with a template](https://support.terra.bio/hc/en-us/articles/360059242671). see \"Sets of data - sample_set table\"\n",
    "\n",
    "- createFailedSampleDataSet.ipynb\n",
    "    The main diffiernece is GTEx data is in a single workspace. TCGA is in multiple"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# expected set format\n",
    "- ID (first column) \n",
    "    The first column is the ID column, the unique name of the set. The format is '<span style=\"color:red\">membership:</span>your-entity-name<span style=\"color:red\">_set_id</span>'. The parts in red are required exactly as typed. The entity-name is whatever the entity you are grouping together. For the Data-QuickStart it will be specimens, but if your workflow processes samples, it could be samples. \n",
    "\n",
    "\n",
    "- Entity column\n",
    "    The second column is the entity you're grouping into sets. The header must match the first column header of the table your workflow will take its single inputs from.\n",
    "\n",
    "Given a sample.tsv file  our sample_set file is \n",
    "```\n",
    "(base) $ cat sample_set_membership.tsv \n",
    "membership:sample_set_id\tsample\n",
    "set-d-4-panc\tGTEX-11GSP-0426-SM-5A5KX\n",
    "set-d-4-panc\tGTEX-11I78-0626-SM-5A5LZ\n",
    "set-d-4-panc\tGTEX-11LCK-0226-SM-5A5M6\n",
    "set-d-4-panc\tGTEX-11NSD-0526-SM-5A5LT\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "run on 2022-03-14 14:20:52\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "now = datetime.now()\n",
    "today = now.strftime('%Y-%m-%d')\n",
    "currentTime = now.strftime('%H:%M:%S')\n",
    "\n",
    "# put time stamp in name. prevents data bug if we accidently run\n",
    "# multiple times. with out time stamp the number of sample displayed\n",
    "# by terra will be a multiple of the true number of samples\n",
    "print(\"run on {}\".format( today +  \" \" + currentTime ))\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'failedPair-2022-03-14-14-20'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set names can not contain ':'\n",
    "newDataSetName = \"failedPair-\" + today + \"-\" + now.strftime('%H-%M') \n",
    "newDataSetName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# back ups of terra data models are stored in a separate repo\n",
    "# so that branch merges do not loose data model version\n",
    "rootDir = \"../../../terraDataModels/test-aedavids-proj/TCGA\"\n",
    "listOfWorkSpacePath = rootDir + \"/\" + \"listOfWorkSpaces.csv\" \n",
    "workspaceDF = pd.read_csv( listOfWorkSpacePath )\n",
    "workSpaceNamesList = workspaceDF.loc[:, \"wokspace\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def readDataModel( rootDir, workspaceName, entityName ) :\n",
    "    dataModelTSV = rootDir + \"/\" + workspaceName + \"/\" + entityName + \".tsv\"\n",
    "    dataModelDF = pd.read_csv(dataModelTSV, delimiter='\\t')\n",
    "    return dataModelDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def saveDataModel( rootDir, workspaceName, entityName, dataModelDF ) :\n",
    "    dataModelTSV = rootDir + \"/\" + workspaceName + \"/\" + entityName + \".tsv\"\n",
    "    print(\"writing {}\".format(dataModelTSV))\n",
    "    dataModelDF.to_csv(dataModelTSV, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "workspace: TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 187\n",
      "num passed:185\n",
      "num failed:2\n",
      "failedSamplesDF.shape:(2, 38)\n"
     ]
    }
   ],
   "source": [
    "def findSamplesWithMissingQuantFile(rootDir, workspaceName, quantFileColName):\n",
    "    sampleDF = readDataModel( rootDir, workspaceName, entityName = \"sample\" )     \n",
    "    print(\"\\nworkspace: {}\".format(workspaceName))\n",
    "    \n",
    "    mRNARowsLogicalPS = sampleDF['mRNASeq_fastq_path'].notna()\n",
    "    nunMRNAFiles = sum( mRNARowsLogicalPS )\n",
    "    print(\"number of mRNASeq_fastq_path files: {}\".format(nunMRNAFiles))\n",
    "    mRNA_DF = sampleDF.loc[mRNARowsLogicalPS,:]\n",
    "\n",
    "    # find rows that have fastq files but are missing results. ie 'quantFilePaired' value\n",
    "    passedSamplesLogicalPS = mRNA_DF[quantFileColName].notna()\n",
    "    numPassed = sum(passedSamplesLogicalPS) \n",
    "    \n",
    "    passedSamplesDF = mRNA_DF.loc[passedSamplesLogicalPS,:]\n",
    "    \n",
    "    failedSampleLogicalPS =  mRNA_DF[quantFileColName].isna()\n",
    "    failedSampleDF = mRNA_DF.loc[failedSampleLogicalPS,:]\n",
    "    \n",
    "    numFailed = nunMRNAFiles - numPassed\n",
    "   \n",
    "    print(\"num passed:{}\".format(numPassed))\n",
    "    print(\"num failed:{}\".format(numFailed))\n",
    "    \n",
    "    return failedSampleDF\n",
    "    \n",
    "\n",
    "def testFindSamplesWithMissingQuantFile():\n",
    "\n",
    "    # 92 subjects in data set\n",
    "    # last job, 184 workflows 79 passed, 105 failed\n",
    "    #\n",
    "#     workspaceName = 'TCGA_ACC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab'\n",
    "#     quantColName  = \"quantFilePaired3\"\n",
    "\n",
    "    #\n",
    "#     workspaceName = \"TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\"\n",
    "#     quantColName  = \"quantFilePaired3\"\n",
    "\n",
    "    workspaceName = \"TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\"\n",
    "    quantColName = \"quantFile3\"\n",
    "    \n",
    "    failedSamplesDF = findSamplesWithMissingQuantFile(rootDir, workspaceName, quantColName)\n",
    "    print(\"failedSamplesDF.shape:{}\".format(failedSamplesDF.shape))\n",
    "    #print(failedSamplesDF)\n",
    "    \n",
    "testFindSamplesWithMissingQuantFile()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def createFailedDataSet( rootDir, workspaceName, failedSamplesDF, newDataSetName ):\n",
    "    sampleSetMembershipDF = readDataModel( rootDir, workspaceName, entityName = \"sample_set_membership\" ) \n",
    "    \n",
    "    ds = [newDataSetName] * failedSamplesDF.shape[0]\n",
    "    # create a copy to eliminate 'SettingWithCopyException' warning\n",
    "    foo = failedSamplesDF.copy()\n",
    "    foo.loc[:, \"membership:sample_set_id\"] = ds\n",
    "    failedSamplesDF = foo\n",
    "    failedSamplesDF.loc[:, \"membership:sample_set_id\"] = ds\n",
    "    \n",
    "    newMembershipDF = failedSamplesDF.loc[:,['membership:sample_set_id', 'entity:sample_id']]\n",
    "    newMembershipDF = newMembershipDF.rename( columns={'membership:sample_set_id':'membership:sample_set_id',\n",
    "                             'entity:sample_id':'sample'} )\n",
    "    newMembershipDF = sampleSetMembershipDF.append( newMembershipDF )\n",
    "    \n",
    "    return newMembershipDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def insertNewDataSetName(rootDir, workspaceName, newDataSetName) :\n",
    "    sampleSetEntityDF = readDataModel( rootDir, workspaceName, entityName = \"sample_set_entity\" )\n",
    "\n",
    "    newSetDF = pd.DataFrame( {\"entity:sample_set_id\":[newDataSetName] })\n",
    "    newSampleSetEntityDF = sampleSetEntityDF.append( newSetDF )\n",
    "\n",
    "    return newSampleSetEntityDF    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def run( rootDir, workspaceName, newDataSetName, quantFileColName ) :\n",
    "    failedSamplesDF = findSamplesWithMissingQuantFile(rootDir, workspaceName, quantFileColName)\n",
    "    if failedSamplesDF.empty :  # or (failedSampleDF.shape[0] == 0)\n",
    "        return\n",
    "    \n",
    "    newMembershipDF      = createFailedDataSet( rootDir, workspaceName, failedSamplesDF, newDataSetName )\n",
    "    newSampleSetEntityDF = insertNewDataSetName(rootDir, workspaceName, newDataSetName)\n",
    "    \n",
    "    saveDataModel(rootDir, workspaceName, \"sample_set_entity\",     newSampleSetEntityDF)\n",
    "    saveDataModel(rootDir, workspaceName, \"sample_set_membership\", newMembershipDF)\n",
    "    \n",
    "    numFailed = failedSamplesDF.shape[0]\n",
    "    return (workspaceName, numFailed)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## figure out which column the results generated by salmonTarQuantWorkflow v 4\n",
    "This workflow uses the salmon paired read bug fix. The reson the name is not always the the same\n",
    "is that there are 33 different workspaces. I had to name the output col manual for each run. For unknow reason\n",
    "Terra would not let define these values in json\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def findQuantFileColumnName() :\n",
    "    dataDict = dict()\n",
    "\n",
    "    missingList = ['TCGA_DLBC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab'\n",
    "                   ,'TCGA_GBM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab'\n",
    "                   ,'TCGA_LAML_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab'\n",
    "                   ,'TCGA_SARC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab'\n",
    "                  ]\n",
    "\n",
    "    for workspaceName in workSpaceNamesList:\n",
    "        if workspaceName in missingList:\n",
    "            continue\n",
    "\n",
    "        #print(workspaceName)\n",
    "        df = readDataModel( rootDir, workspaceName, entityName = \"sample\" )\n",
    "        quantMatchlist = [s for s in df.columns if \"quantF\" in s]\n",
    "        colName = [s for s in quantMatchlist if \"3\" in s][0]\n",
    "        #print(colName)\n",
    "        #print()\n",
    "        dataDict[workspaceName] = (colName, df) \n",
    "        \n",
    "    return dataDict\n",
    "        \n",
    "dataDict = findQuantFileColumnName()        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "workspace: TCGA_ACC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 79\n",
      "num passed:79\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 40)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_BLCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 427\n",
      "num passed:427\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 56)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_BRCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 1215\n",
      "num passed:1181\n",
      "num failed:34\n",
      "failedSampleDF.shape:(34, 57)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_CESC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 309\n",
      "num passed:309\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 48)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_CHOL_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 45\n",
      "num passed:45\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 40)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_COAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 501\n",
      "num passed:309\n",
      "num failed:192\n",
      "failedSampleDF.shape:(192, 64)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_ESCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 196\n",
      "num passed:196\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 49)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_HNSC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 566\n",
      "num passed:566\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 56)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_KICH_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 91\n",
      "num passed:91\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 48)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_KIRC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 606\n",
      "num passed:606\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 58)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_KIRP_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 323\n",
      "num passed:323\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 53)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_LGG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 530\n",
      "num passed:530\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 51)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_LIHC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 423\n",
      "num passed:423\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 49)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 576\n",
      "num passed:535\n",
      "num failed:41\n",
      "failedSampleDF.shape:(41, 61)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_LUSC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 552\n",
      "num passed:552\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 61)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_MESO_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 87\n",
      "num passed:87\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 36)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_OV_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 298\n",
      "num passed:298\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 57)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 183\n",
      "num passed:183\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 42)\n",
      "quantFilePaired3\n",
      "\n",
      "\n",
      "workspace: TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 187\n",
      "num passed:185\n",
      "num failed:2\n",
      "failedSampleDF.shape:(2, 38)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_PRAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 550\n",
      "num passed:535\n",
      "num failed:15\n",
      "failedSampleDF.shape:(15, 44)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_READ_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 177\n",
      "num passed:105\n",
      "num failed:72\n",
      "failedSampleDF.shape:(72, 63)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_SKCM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 473\n",
      "num passed:473\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 48)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_STAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 450\n",
      "num passed:411\n",
      "num failed:39\n",
      "failedSampleDF.shape:(39, 60)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_TGCT_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 156\n",
      "num passed:156\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 40)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_THCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 568\n",
      "num passed:564\n",
      "num failed:4\n",
      "failedSampleDF.shape:(4, 50)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_THYM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 122\n",
      "num passed:122\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 39)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_UCEC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 581\n",
      "num passed:201\n",
      "num failed:380\n",
      "failedSampleDF.shape:(380, 65)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_UCS_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 57\n",
      "num passed:57\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 41)\n",
      "quantFile3\n",
      "\n",
      "\n",
      "workspace: TCGA_UVM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 80\n",
      "num passed:80\n",
      "num failed:0\n",
      "failedSampleDF.shape:(0, 44)\n",
      "quantFile3\n"
     ]
    }
   ],
   "source": [
    "for workspaceName,t in dataDict.items():\n",
    "    print()\n",
    "#     print(workspaceName)\n",
    "    colName, df = t\n",
    "    failedSampleDF = findSamplesWithMissingQuantFile(rootDir, workspaceName, colName)\n",
    "    print(\"failedSampleDF.shape:{}\".format(failedSampleDF.shape))\n",
    "    print(colName)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## explore TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
    "Exploration of ESCA, OV, and STAD data set failures revealed that the mRNASeq_fastq_path files where really tar files containing paired fastq file. These 3 data sets had 100% failures when we ran salmonSingleEndReadTask.wdl. This leads to the possiblity that other TCGA data sets might also contain paired fastq files. As a test we ran salmonTarQuantWorkflow.wdl on 2/11. Compare results from quantFile and quantFilePaired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "workspaceName = \"TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\"\n",
    "# failedSamplesDF      = findSamplesWithMissingQuantFile(rootDir, workspaceName)\n",
    "# failedSamplesDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "missing single and paired results\n",
      "             entity:sample_id  quantFile  quantFilePaired\n",
      "sample_type                                              \n",
      "NB                        153          0                0\n",
      "NT                         33          0                0\n",
      "TP                          7          0                0\n",
      "\n",
      " missing single or paired results\n",
      "             entity:sample_id  quantFile  quantFilePaired\n",
      "sample_type                                              \n",
      "NB                        153          0                0\n",
      "NT                         33          0                0\n",
      "TP                          8          0                1\n",
      "\n",
      "type == TP and either quantFile or quantFilePaired is missing\n",
      "    entity:sample_id sample_type  quantFile.isna  quantFilePaired.isna\n",
      "98   PAAD-F2-A8YN-TP          TP            True                 False\n",
      "134  PAAD-FZ-5919-TP          TP            True                  True\n",
      "136  PAAD-FZ-5920-TP          TP            True                  True\n",
      "138  PAAD-FZ-5921-TP          TP            True                  True\n",
      "140  PAAD-FZ-5922-TP          TP            True                  True\n",
      "142  PAAD-FZ-5923-TP          TP            True                  True\n",
      "144  PAAD-FZ-5924-TP          TP            True                  True\n",
      "146  PAAD-FZ-5926-TP          TP            True                  True\n",
      "\n",
      "\n",
      " are these truely single end reads?\n",
      "Empty DataFrame\n",
      "Columns: [entity:sample_id, sample_type, quantFile.isna, quantFilePaired.isna]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "def explore( workspaceName ) :\n",
    "    sampleDF = readDataModel( rootDir, workspaceName, entityName = \"sample\" ) \n",
    "    colList = ['entity:sample_id', 'sample_type','quantFile', 'quantFilePaired', 'aux_info', 'aux_infoPaired']\n",
    "    #print( sampleDF.loc[:, colList].head() )\n",
    "    \n",
    "    missingQuantFileRows = sampleDF['quantFile'].isna()\n",
    "    missingQuantFilePairedRows = sampleDF['quantFilePaired'].isna()\n",
    "    hasMRnaSeqRows = sampleDF['mRNASeq_fastq_path'].notna()\n",
    "    \n",
    "    missingSingleAndPairedResultsDF = sampleDF.loc[ missingQuantFileRows & missingQuantFilePairedRows, colList]\n",
    "    #print(\"missing singale and paired results\\n{}\".format(missingSingleAndPairedResultsDF.loc[:, colList]) )\n",
    "    print(\"missing single and paired results\")\n",
    "    cl = ['entity:sample_id','quantFile', 'quantFilePaired','sample_type']\n",
    "    print( missingSingleAndPairedResultsDF.loc[:, cl].groupby('sample_type').count() )\n",
    "    \n",
    "    missingSingleOrPairedResultsDF = sampleDF.loc[ missingQuantFileRows | missingQuantFilePairedRows, colList]\n",
    "    print(\"\\n missing single or paired results\")\n",
    "    print( missingSingleOrPairedResultsDF.loc[:,cl].groupby('sample_type').count() )   \n",
    "    \n",
    "    tpRows = sampleDF['sample_type'] == \"TP\"\n",
    " \n",
    "    printableDF = sampleDF.loc[(tpRows & (missingQuantFileRows | missingQuantFilePairedRows)), cl]\n",
    "    printableDF.loc[:,\"quantFile.isna\"] = printableDF.loc[:,\"quantFile\"].isna()\n",
    "    printableDF.loc[:,\"quantFilePaired.isna\"] = printableDF.loc[:,\"quantFilePaired\"].isna()\n",
    "    print(\"\\ntype == TP and either quantFile or quantFilePaired is missing\")\n",
    "    print(printableDF.loc[:, ['entity:sample_id', 'sample_type', 'quantFile.isna', 'quantFilePaired.isna']])\n",
    "    \n",
    "    print(\"\\n\\n are these truely single end reads?\")\n",
    "    r = (~ printableDF.loc[:, 'quantFile.isna']) & printableDF.loc[:, 'quantFilePaired.isna'] \n",
    "    print( printableDF.loc[r, ['entity:sample_id', 'sample_type', 'quantFile.isna', 'quantFilePaired.isna']])\n",
    "    \n",
    "          \n",
    "explore( workspaceName )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### explore job results\n",
    "https://app.terra.bio/#workspaces/test-aedavids-proj/TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/job_history/922ad136-e8ad-43f5-b946-5669a5a16d04\n",
    "\n",
    "all 4 sample (PAAD-F2-7273-TP, PAAD-HZ-7924-TP, PAAD-HZ-8519-TP, PAAD-IB-8127-TP ) failed the same way\n",
    "\n",
    "```\n",
    "Task quantify.tarToFastqTask:NA:1 failed\n",
    "\n",
    "while running \"-c grep -E -q 'OutOfMemory|Killed' /cromwell_root/stderr ; echo $? > /cromwell_root/memory_retry_rc\": unexpected exit status 1 was not ignored [CheckingForMemoryRetry] Unexpected exit status 1 while running \"-c grep -E -q 'OutOfMemory|Killed' /cromwell_root/stderr ; echo $? > /cromwell_root/memory_retry_rc\": sh: write error: No space left on device\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## run all workspaces\n",
    "<span style=\"color:red\">check out missingList in findQuantFileColumnName. we do not test all workspaces</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "workspace: TCGA_ACC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 79\n",
      "num passed:79\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_BLCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 427\n",
      "num passed:427\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_BRCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 1215\n",
      "num passed:1181\n",
      "num failed:34\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_BRCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_BRCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_CESC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 309\n",
      "num passed:309\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_CHOL_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 45\n",
      "num passed:45\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_COAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 501\n",
      "num passed:309\n",
      "num failed:192\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_COAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_COAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_ESCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 196\n",
      "num passed:196\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_HNSC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 566\n",
      "num passed:566\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_KICH_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 91\n",
      "num passed:91\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_KIRC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 606\n",
      "num passed:606\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_KIRP_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 323\n",
      "num passed:323\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_LGG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 530\n",
      "num passed:530\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_LIHC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 423\n",
      "num passed:423\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 576\n",
      "num passed:535\n",
      "num failed:41\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_LUSC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 552\n",
      "num passed:552\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_MESO_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 87\n",
      "num passed:87\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_OV_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 298\n",
      "num passed:298\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_PAAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 183\n",
      "num passed:183\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 187\n",
      "num passed:185\n",
      "num failed:2\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_PRAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 550\n",
      "num passed:535\n",
      "num failed:15\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_PRAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_PRAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_READ_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 177\n",
      "num passed:105\n",
      "num failed:72\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_READ_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_READ_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_SKCM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 473\n",
      "num passed:473\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_STAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 450\n",
      "num passed:411\n",
      "num failed:39\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_STAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_STAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_TGCT_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 156\n",
      "num passed:156\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_THCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 568\n",
      "num passed:564\n",
      "num failed:4\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_THCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_THCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_THYM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 122\n",
      "num passed:122\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_UCEC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 581\n",
      "num passed:201\n",
      "num failed:380\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_UCEC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_entity.tsv\n",
      "writing ../../../terraDataModels/test-aedavids-proj/TCGA/TCGA_UCEC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab/sample_set_membership.tsv\n",
      "\n",
      "workspace: TCGA_UCS_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 57\n",
      "num passed:57\n",
      "num failed:0\n",
      "\n",
      "workspace: TCGA_UVM_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\n",
      "number of mRNASeq_fastq_path files: 80\n",
      "num passed:80\n",
      "num failed:0\n",
      "\n",
      "\n",
      " set name:failedPair-2022-03-14-14-20 workspaces with failed samples\n",
      "('TCGA_BRCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 34)\n",
      "('TCGA_COAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 192)\n",
      "('TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 41)\n",
      "('TCGA_PCPG_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 2)\n",
      "('TCGA_PRAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 15)\n",
      "('TCGA_READ_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 72)\n",
      "('TCGA_STAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 39)\n",
      "('TCGA_THCA_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 4)\n",
      "('TCGA_UCEC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab', 380)\n",
      "check out missingList in findQuantFileColumnName. we do not test all workspaces\n"
     ]
    }
   ],
   "source": [
    "dataDict = findQuantFileColumnName()        \n",
    "workspacesWithFailedSamples = list()\n",
    "\n",
    "# for workspaceName in workSpaceNamesList :\n",
    "# bugFixWS = [\"TCGA_SARC_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\", \n",
    "#             \"TCGA_LUAD_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\"]\n",
    "# for workspaceName in bugFixWS :\n",
    "\n",
    "for workspaceName,t in dataDict.items():\n",
    "    colName, df = t\n",
    "    try:\n",
    "        failed_ws = run( rootDir, workspaceName, newDataSetName, colName )\n",
    "        if failed_ws != None:\n",
    "            workspacesWithFailedSamples.append( failed_ws )\n",
    "    except Exception as e:\n",
    "        print(\"\\n\\nERROR: {} {}\\n\\n\".format(workspaceName, e))\n",
    "\n",
    "print(\"\\n\\n set name:{} workspaces with failed samples\".format(newDataSetName))\n",
    "for ws in workspacesWithFailedSamples:\n",
    "    print( ws )\n",
    "print(\"check out missingList in findQuantFileColumnName. we do not test all workspaces\")\n",
    "\n",
    "# workspaceName = \"TCGA_KICH_ControlledAccess_V1-0_DATA_edu_ucsc_kim_lab\"\n",
    "# run( rootDir, workspaceName, newDataSetName )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:extraCellularRNA] *",
   "language": "python",
   "name": "conda-env-extraCellularRNA-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
