{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ed0d5968-7404-47e9-af4a-2e46f0f2a53d",
   "metadata": {},
   "source": [
    "# Random Forest GTEx_TCGA Gene Signature \n",
    "\n",
    "Andrew E. Davidson  \n",
    "aedavids@ucsc.edu  \n",
    "7/15/24\n",
    "\n",
    "Copyright (c) 2020-2023, Regents of the University of California All rights reserved. https://polyformproject.org/licenses/noncommercial/1.0.0\n",
    "\n",
    "ref:  \n",
    "* extraCellularRNA/deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolution.ipynb\n",
    "* extraCellularRNA/deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolutionPOC.ipynb\n",
    "* extraCellularRNA/intraExtraRNA_POC/jupyterNotebooks/elife/elifeBinaryRandomForestResults.ipynb\n",
    "\n",
    "## <span style=\"color:red;background-color:yellow\">TODO</span>\n",
    "- deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolution.ipynb\n",
    "\n",
    "- deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolutionPOC.ipynb\n",
    "\n",
    "  \n",
    "- deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolutionPOC.ipynb\n",
    "  * can we find these missing genes\n",
    "  * KeyError: \"['ENSG00000263264.2', 'ENSG00000288380.1', 'ENSG00000274031.1'] not in index\"\n",
    "\n",
    "\n",
    "## **<span style=\"color:red;background-color:yellow\">model bulk tissue models fails to predict elife</span>**  \n",
    "92% are predicted to be 'Whole_Blood'.\n",
    "\n",
    "- todo\n",
    "  * treat like mulit model. ie what is the next most probably class\n",
    "  * train a new model on bulk wiht out whole_blood bio markers. Label whole_blood samples as unknown. I.E. not one of the 82 other classes\n",
    "    + elife healthy control should map to unknown?\n",
    "  * look for likelihood vectors with no strong signal "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07bdb288-f1cb-4d95-aad7-8493b5eeb7d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/tmp/ipykernel_2495527/3098827269.py:11: DeprecationWarning: \n",
      "Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),\n",
      "(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)\n",
      "but was not found to be installed on your system.\n",
      "If this would cause problems for you,\n",
      "please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466\n",
      "        \n",
      "  import pandas as pd\n"
     ]
    }
   ],
   "source": [
    "import ipynbname\n",
    "\n",
    "# use display() to print an html version of a data frame\n",
    "# useful if dataFrame output is not generated by last like of cell\n",
    "from IPython.display import display\n",
    "import joblib\n",
    "import logging\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "# display all columns\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)\n",
    "\n",
    "# from sklearn.metrics import confusion_matrix\n",
    "from sklearn.ensemble      import RandomForestClassifier\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "# from sklearn.metrics         import recall_score\n",
    "# from sklearn.metrics         import roc_auc_score\n",
    "# from sklearn.metrics         import make_scorer\n",
    "\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b472d3f-1383-4f4e-8858-bdd73c5589c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outDir:\n",
      "/private/groups/kimlab/aedavids/deconvolution/randomForestGTEx_TCGAGeneSignature.out\n",
      "\n",
      "modelOutDir ;\n",
      "/private/groups/kimlab/aedavids/deconvolution/randomForestGTEx_TCGAGeneSignature.out/model\n",
      "\n",
      "imgOut :\n",
      "/private/groups/kimlab/aedavids/deconvolution/randomForestGTEx_TCGAGeneSignature.out/img\n"
     ]
    }
   ],
   "source": [
    "notebookName = ipynbname.name()\n",
    "notebookPath = ipynbname.path()\n",
    "notebookDir = os.path.dirname(notebookPath)\n",
    "\n",
    "#outDir = f'{notebookDir}/{notebookName}.out'\n",
    "outDir = f'/private/groups/kimlab/aedavids/deconvolution/{notebookName}.out'\n",
    "os.makedirs(outDir, exist_ok=True)\n",
    "print(f'outDir:\\n{outDir}')\n",
    "\n",
    "modelOutDir = os.path.join(outDir, \"model\")\n",
    "os.makedirs(modelOutDir, exist_ok=True)\n",
    "print(f'\\nmodelOutDir ;\\n{modelOutDir}')\n",
    "\n",
    "imgOut = f'{outDir}/img'\n",
    "os.makedirs(imgOut, exist_ok=True)\n",
    "print(f'\\nimgOut :\\n{imgOut}')\n",
    "\n",
    "#loglevel = \"INFO\"\n",
    "loglevel = \"WARN\"\n",
    "# logFMT = \"%(asctime)s %(levelname)s [thr:%(threadName)s %(name)s %(funcName)s() line:%(lineno)s] [%(message)s]\"\n",
    "logFMT = \"%(asctime)s %(levelname)s %(name)s %(funcName)s() line:%(lineno)s] [%(message)s]\"\n",
    "logging.basicConfig(format=logFMT, level=loglevel)    \n",
    "logger = logging.getLogger(notebookName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae5fdf95-8468-463f-9600-6958b74e91d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORIG_PYTHONPATH: :/private/home/aedavids/extraCellularRNA/src\n",
      "\n",
      "deconvolutionModules: /private/home/aedavids/extraCellularRNA/deconvolutionAnalysis/python\n",
      "\n",
      "intraExtraRNA_POCModules: /private/home/aedavids/extraCellularRNA/intraExtraRNA_POC/python/src\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# setting the python path allows us to run python scripts from using\n",
    "# the CLI. \n",
    "PYTHONPATH = os.environ['PYTHONPATH']\n",
    "print(\"ORIG_PYTHONPATH: {}\\n\".format(PYTHONPATH))\n",
    "\n",
    "gitRepoRoot = !git rev-parse --show-toplevel\n",
    "gitRepoRoot = gitRepoRoot[0]\n",
    "\n",
    "#\n",
    "# add deconvolutionAnalysis modules\n",
    "#\n",
    "deconvolutionModules = f'{gitRepoRoot}/deconvolutionAnalysis/python'\n",
    "print(\"deconvolutionModules: {}\\n\".format(deconvolutionModules))\n",
    "\n",
    "PYTHONPATH = PYTHONPATH + f':{deconvolutionModules}'\n",
    "#print(\"PYTHONPATH: {}\\n\".format(PYTHONPATH))\n",
    "\n",
    "sys.path.append( str(deconvolutionModules) )\n",
    "#print(\"\\nsys.path:\\n{}\\n\".format(sys.path))\n",
    "\n",
    "#\n",
    "# add intraExtraRNA_POC modules\n",
    "#\n",
    "intraExtraRNA_POCModules = f'{gitRepoRoot}/intraExtraRNA_POC/python/src'\n",
    "print(\"intraExtraRNA_POCModules: {}\\n\".format(intraExtraRNA_POCModules))\n",
    "\n",
    "PYTHONPATH = PYTHONPATH + f':{intraExtraRNA_POCModules}'\n",
    "#print(\"PYTHONPATH: {}\\n\".format(PYTHONPATH))\n",
    "\n",
    "sys.path.append( str(intraExtraRNA_POCModules) )\n",
    "#print(\"\\nsys.path:\\n{}\\n\".format(sys.path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0364cd5f-422c-4be8-b1c8-31eb5c6f244b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# local imports\n",
    "# from analysis.utilities import saveList\n",
    "# from analysis.utilities import loadList\n",
    "from intraExtraRNA.elifeUtilities import loadElifeTrainingData\n",
    "from intraExtraRNA.elifeUtilities import validElifeCategories\n",
    "from models.mlUtilities import encoder2Dict\n",
    "from models.mlUtilities import loadEncoder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68edf461-12a1-421d-9a21-a5de8bac109b",
   "metadata": {},
   "source": [
    "## Load Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d4c5599e-f7a8-47c0-8f05-7c05f42680ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "colDataDF.shape : (15801, 6)\n",
      "len(categories) : 83\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Adipose_Subcutaneous', 'Artery_Tibial', 'Heart_Atrial_Appendage',\n",
       "       'Breast_Mammary_Tissue', 'Brain_Cortex'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a list of all the GTEx and TCGA types and classes\n",
    "colDataPath = \"/private/groups/kimlab/GTEx_TCGA/groupbyGeneTrainingSets/GTEx_TCGA_TrainColData.csv\"\n",
    "colDataDF = pd.read_csv(colDataPath)\n",
    "print(f'colDataDF.shape : {colDataDF.shape}')\n",
    "#colDataDF.head()\n",
    "categories = colDataDF.loc[:, \"category\"].unique()\n",
    "print(f'len(categories) : {len(categories)}')\n",
    "categories[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ddc2142d-0650-4ce2-a401-b81e694a80f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-24 13:53:03,408 WARNING intraExtraRNA.elifeUtilities selectFeatures() line:417] [missingInV39Set: {'ENSG00000288380', 'ENSG00000274031', 'LTR106', 'ENSG00000263264', 'ENSG00000244693'}]\n",
      "2024-07-24 13:53:03,410 WARNING intraExtraRNA.elifeUtilities selectFeatures() line:433] [missingElifeGenes: set()]\n",
      "2024-07-24 13:53:03,412 WARNING intraExtraRNA.elifeUtilities selectFeatures() line:433] [missingElifeGenes: set()]\n",
      "2024-07-24 13:53:03,412 WARNING intraExtraRNA.elifeUtilities selectFeatures() line:433] [missingElifeGenes: set()]\n",
      "2024-07-24 13:53:03,413 WARNING intraExtraRNA.elifeUtilities selectFeatures() line:433] [missingElifeGenes: set()]\n",
      "2024-07-24 13:53:03,718 WARNING intraExtraRNA.elifeUtilities fixBest10CuratedDegree1_ce467ff() line:95] [elifeLungGenes replacing ENSG00000182378.15_PAR_Y with ENSG00000182378.15 elifeLungGenes]\n",
      "2024-07-24 13:53:03,719 WARNING intraExtraRNA.elifeUtilities loadElifeTrainingData() line:285] [missingElifeGenes\n",
      " : ['ENSG00000288380', 'ENSG00000274031', 'ENSG00000263264', 'ENSG00000244693']]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "missingGenes:\n",
      "['ENSG00000288380', 'ENSG00000274031', 'ENSG00000263264', 'ENSG00000244693']\n",
      "CPU times: user 1min 19s, sys: 12.5 s, total: 1min 31s\n",
      "Wall time: 1min 33s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "pipelineStageName = \"best10CuratedDegree1_ce467ff\"\n",
    "t = loadElifeTrainingData(pipelineStageName,\n",
    "                             categories,\n",
    "                             validElifeCategories,\n",
    "                             )\n",
    "\n",
    "HUGO_Genes, elifeGenes, missingGenes, countDF, metaDF, XDF, yNP, labelEncoderElife, mapDF = t\n",
    "\n",
    "print(f'missingGenes:\\n{missingGenes}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f041e26c-7acb-48c0-8b2a-f8f4a0451423",
   "metadata": {},
   "source": [
    "## Load model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5e8dac39-ecbe-4a39-a6f3-74b685404366",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading model: /private/groups/kimlab/aedavids/deconvolution/randomForestGeneSignatureDeconvolution.out/model/best10CuratedDegree1_ce467ff_elife.joblib\n",
      "modelName : best10CuratedDegree1_ce467ff_elife\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;RandomForestClassifier<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(random_state=42)</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier(random_state=42)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#modelName = 'best10CuratedDegree1_ce467ff' # full model with 716 bio marker genes\n",
    "modelName = 'best10CuratedDegree1_ce467ff_elife' # does not include 4 missing genes \n",
    "modelRootDir = '/private/groups/kimlab/aedavids/deconvolution/randomForestGeneSignatureDeconvolution.out/model'\n",
    "modelPath = f\"{modelRootDir}/{modelName}.joblib\"\n",
    "\n",
    "print(f'loading model: {modelPath}')\n",
    "model = joblib.load(modelPath)\n",
    "\n",
    "print(f'modelName : {modelName}')\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dc43bb9-b6ed-4ba7-9a0d-69d8aa64b4d0",
   "metadata": {},
   "source": [
    "### find missing features\n",
    "The full best10CuratedDegree1_ce467ff run used v35 genes named using HUGO ids. The full random forest has 716 genes. \n",
    "Elife uses v39 ENSG ids. 4 v35 genes are missing. Find the missing ids in HUGO format and save them to the model out directory\n",
    "This will make it possible to train a elife specific model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c4da9e55-0866-4dae-906b-b124bac27759",
   "metadata": {},
   "outputs": [],
   "source": [
    "if modelName == 'best10CuratedDegree1_ce467ff':\n",
    "    # find mapping for features in elife that match our model\n",
    "    hugoModelFeatures = model.feature_names_in_\n",
    "    selectHugoModelFeature = mapDF.loc[:, \"HUGO_v35\"].isin( hugoModelFeatures )\n",
    "    \n",
    "    matchingMapDF = mapDF.loc[ selectHugoModelFeature, : ]\n",
    "    print( f'number of features : {model.n_features_in_}' )\n",
    "    print(f'matchingMapDF.shape : {matchingMapDF.shape}' )\n",
    "    display(matchingMapDF.head())\n",
    "    \n",
    "    # find model features that are not in elife in Hugo format\n",
    "    missingElifeGenesInHUGOFormat = list( set( hugoModelFeatures ) - set( matchingMapDF.loc[:, 'HUGO_v35'] ) )\n",
    "    print(f'missing elife genes in hugo format : {missingElifeGenesInHUGOFormat}' )\n",
    "\n",
    "    missingElifeFeaturesPath =  f\"{modelRootDir}/missingElifeFeatures.csv\"\n",
    "    missingElifeGenesSeries = pd.Series( missingElifeGenesInHUGOFormat, name=\"missing\" )\n",
    "    missingElifeGenesSeries.to_csv( missingElifeFeaturesPath, index=False )\n",
    "    print(f'saving missing features list to :\\n{missingElifeFeaturesPath}' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b0b4b0ef-f265-4abf-92fe-111cd05617f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labelEncoder: \n",
      "{'ACC': 0, 'Adipose_Subcutaneous': 1, 'Adipose_Visceral_Omentum': 2, 'Adrenal_Gland': 3, 'Artery_Aorta': 4, 'Artery_Coronary': 5, 'Artery_Tibial': 6, 'BLCA': 7, 'BRCA': 8, 'Bladder': 9, 'Brain_Amygdala': 10, 'Brain_Anterior_cingulate_cortex_BA24': 11, 'Brain_Caudate_basal_ganglia': 12, 'Brain_Cerebellar_Hemisphere': 13, 'Brain_Cerebellum': 14, 'Brain_Cortex': 15, 'Brain_Frontal_Cortex_BA9': 16, 'Brain_Hippocampus': 17, 'Brain_Hypothalamus': 18, 'Brain_Nucleus_accumbens_basal_ganglia': 19, 'Brain_Putamen_basal_ganglia': 20, 'Brain_Spinal_cord_cervical_c-1': 21, 'Brain_Substantia_nigra': 22, 'Breast_Mammary_Tissue': 23, 'CESC': 24, 'CHOL': 25, 'COAD': 26, 'Cells_Cultured_fibroblasts': 27, 'Cells_EBV-transformed_lymphocytes': 28, 'Cervix_Endocervix': 29, 'Colon_Sigmoid': 30, 'Colon_Transverse': 31, 'DLBC': 32, 'ESCA': 33, 'Esophagus_Gastroesophageal_Junction': 34, 'Esophagus_Mucosa': 35, 'Esophagus_Muscularis': 36, 'GBM': 37, 'HNSC': 38, 'Heart_Atrial_Appendage': 39, 'Heart_Left_Ventricle': 40, 'KICH': 41, 'KIRC': 42, 'KIRP': 43, 'Kidney_Cortex': 44, 'LGG': 45, 'LIHC': 46, 'LUAD': 47, 'LUSC': 48, 'Liver': 49, 'Lung': 50, 'MESO': 51, 'Minor_Salivary_Gland': 52, 'Muscle_Skeletal': 53, 'Nerve_Tibial': 54, 'OV': 55, 'Ovary': 56, 'PAAD': 57, 'PCPG': 58, 'PRAD': 59, 'Pancreas': 60, 'Pituitary': 61, 'Prostate': 62, 'READ': 63, 'SARC': 64, 'SKCM': 65, 'STAD': 66, 'Skin_Not_Sun_Exposed_Suprapubic': 67, 'Skin_Sun_Exposed_Lower_leg': 68, 'Small_Intestine_Terminal_Ileum': 69, 'Spleen': 70, 'Stomach': 71, 'TGCT': 72, 'THCA': 73, 'THYM': 74, 'Testis': 75, 'Thyroid': 76, 'UCEC': 77, 'UCS': 78, 'UVM': 79, 'Uterus': 80, 'Vagina': 81, 'Whole_Blood': 82}\n",
      "labelEncoder len(classes) :\n",
      "83\n"
     ]
    }
   ],
   "source": [
    "encoderPath = f'{modelRootDir}/{modelName}.labelEncoder.txt'\n",
    "labelEncoder = loadEncoder(encoderPath)\n",
    "print(f'labelEncoder: \\n{encoder2Dict(labelEncoder)}')\n",
    "\n",
    "classes = labelEncoder.classes_\n",
    "print(f'labelEncoder len(classes) :\\n{len(classes)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b972b9fc-93c9-44a0-aa30-36c5f2637dd6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "54eb19cd-77d8-45d8-956f-2d79df645edf",
   "metadata": {},
   "source": [
    "## Make Predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33537800-6cdf-4c03-b428-4d761bb6e5b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e0b85b2-b86c-491c-b4ca-27167a552db4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-24 13:53:05,227 ERROR randomForestGTEx_TCGAGeneSignature <module>() line:8] [AEDWIP elife has an extra feature]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "712\n",
      "len(orderedFeatureHugoList) : 712\n",
      "len(XDF.columns) : 713\n"
     ]
    }
   ],
   "source": [
    "# we need to make sure the features (cols) are in correct order\n",
    "print(model.n_features_in_)\n",
    "# The attribute feature_names_in_ will only be valid if model was trained using a data frame\n",
    "# numpy arrays to not have feature name information\n",
    "orderedFeatureHugoList = model.feature_names_in_\n",
    "print( f'len(orderedFeatureHugoList) : {len(orderedFeatureHugoList)}' )\n",
    "print(f'len(XDF.columns) : {len(XDF.columns)}' )\n",
    "logger.error('AEDWIP elife has an extra feature')\n",
    "# print( f'\\norderedFeatureHugoList[0:5] : { orderedFeatureHugoList[0:5]} ') \n",
    "# print( f'\\norderedFeatureHugoList[-5:] : { orderedFeatureHugoList[-5:]} ') \n",
    "\n",
    "# print( 'PLCXD1' in orderedFeatureHugoList ) \n",
    "# # display( mapDF.loc['PLCXD1', :] )\n",
    "# display( mapDF[ mapDF['HUGO_v35'] =='PLCXD1' ] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "655d771a-e839-4fbd-bb80-076300693709",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "begin\n",
      "len(orderedFeatureHugoList) : 712\n",
      "len(XDF.columns) : 713\n",
      "ENSG00000182378.15 is a duplicate idx: 670\n",
      "ENSG00000182378.15 is a duplicate idx: 712\n",
      "[670, 712]\n",
      "713\n",
      "712\n",
      "tmpXDF.shape : (224, 712)\n"
     ]
    }
   ],
   "source": [
    "def dropDuplicateColumns(orderedFeatureHugoList, XDF, mapDF):\n",
    "    '''\n",
    "    TODO\n",
    "\n",
    "    not generic. Assume only 1 key is duplicate\n",
    "\n",
    "    I forgot why there is a duplicate. I think it might be because a given gene\n",
    "    can have several transcripts each with a different bio type. this results\n",
    "    in multiple entries in teh mapping table\n",
    "\n",
    "    could be because of hack I added to map elife genes\n",
    "    '''\n",
    "    print('begin')\n",
    "    print( f'len(orderedFeatureHugoList) : {len(orderedFeatureHugoList)}' )\n",
    "    print(f'len(XDF.columns) : {len(XDF.columns)}' )\n",
    "\n",
    "    #display( list(XDF.columns) )\n",
    "    # find extra feature?\n",
    "    #print( sum(XDF.columns == 'ENSG00000182378.15') )\n",
    "    #display( mapDF[ mapDF['ENSG_v39'] =='ENSG00000182378.15' ] )\n",
    "    \n",
    "    # check for dups\n",
    "    columnsLst = list(XDF.columns)\n",
    "    offset = -1\n",
    "    dupIdxs = []\n",
    "    for c in columnsLst:\n",
    "        if columnsLst.count(c) > 1:\n",
    "            offset =columnsLst.index(c, offset+1 ) \n",
    "            print(f\"{c} is a duplicate idx: {offset}\")\n",
    "            dupIdxs.append( offset )\n",
    "\n",
    "    print(dupIdxs)\n",
    "    \n",
    "\n",
    "    v1 = XDF.iloc[:, dupIdxs[0] ]\n",
    "    v2 = XDF.iloc[:, dupIdxs[1] ]\n",
    "\n",
    "    assert v1.equals(v2) , f'ERROR :The column names are the sames how ever values are differnt'\n",
    "\n",
    "    # drop idx from end of list first\n",
    "    lastDupIdx = dupIdxs[-1]\n",
    "    print( len(columnsLst) )\n",
    "    # del columnsLst[lastDupIdx]\n",
    "    intIdx = [i for i in range(len(columnsLst)) ]\n",
    "    del intIdx[lastDupIdx]\n",
    "    # print( len(columnsLst) )\n",
    "    print( len(intIdx) )\n",
    "\n",
    "    return XDF.iloc[:, intIdx].copy()\n",
    "\n",
    "\n",
    "tmpXDF = dropDuplicateColumns(orderedFeatureHugoList, XDF, mapDF)\n",
    "print(f'tmpXDF.shape : {tmpXDF.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "024c9966-87d9-4b71-bea5-0b971a4c0b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# the values in the mapDF columns are not unique (a gene can have several transcripts, with different biotypes)\n",
    "# do not set index\n",
    "# tmpDF1 = mapDF.copy()\n",
    "# # use tail to debug. the ENSG biomarkers are on the bottom. The top are repeats\n",
    "# display( tmpDF1.tail() )\n",
    "\n",
    "# print()\n",
    "# display( tmpDF1.tail() )\n",
    "\n",
    "# print()\n",
    "# # display(tmpDF1.loc['PLCXD1', :] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "644cf661-c7d0-4f28-9ee6-b9e55e893201",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reorder the rows\n",
    "# V39FeatureOrderSeries = tmpDF1.loc[ orderedFeatureHugoList, 'ENSG_v39' ]\n",
    "# display( V39FeatureOrderSeries.head() )\n",
    "# print()\n",
    "# display( V39FeatureOrderSeries.tail() )\n",
    "# print()\n",
    "# print( V39FeatureOrderSeries['ENSG00000182378.15'] )\n",
    "#print( V39FeatureOrderSeries['ENSG00000182378.15_PAR_Y'] )\n",
    "# print( V39FeatureOrderSeries['ENSG00000182378.14'] )\n",
    "# print( V39FeatureOrderSeries[''] )\n",
    "# print( V39FeatureOrderSeries[''] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6fd0ccd2-778f-46e3-b96a-ee55c995b31f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "we can not use the hugo and elife genes list they are not order correctly\n",
      "['TPRN', 'DENND6B', 'CRTAC1', 'AC069281.2', 'SLC9A3-AS1']\n",
      "['ENSG00000198744.5', 'ENSG00000187642.9', 'ENSG00000215915.10', 'ENSG00000235169.11', 'ENSG00000171819.5']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HUGO_v35</th>\n",
       "      <th>ENSG_v35</th>\n",
       "      <th>ENSG_v39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>SLC9A3-AS1</td>\n",
       "      <td>ENSG00000225138.8</td>\n",
       "      <td>ENSG00000225138.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>AC069281.2</td>\n",
       "      <td>ENSG00000274272.1</td>\n",
       "      <td>ENSG00000274272.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>TPRN</td>\n",
       "      <td>ENSG00000176058.13</td>\n",
       "      <td>ENSG00000176058.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>CRTAC1</td>\n",
       "      <td>ENSG00000095713.14</td>\n",
       "      <td>ENSG00000095713.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>669</th>\n",
       "      <td>DENND6B</td>\n",
       "      <td>ENSG00000205593.12</td>\n",
       "      <td>ENSG00000205593.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       HUGO_v35            ENSG_v35            ENSG_v39\n",
       "174  SLC9A3-AS1   ENSG00000225138.8   ENSG00000225138.8\n",
       "257  AC069281.2   ENSG00000274272.1   ENSG00000274272.1\n",
       "320        TPRN  ENSG00000176058.13  ENSG00000176058.13\n",
       "337      CRTAC1  ENSG00000095713.14  ENSG00000095713.14\n",
       "669     DENND6B  ENSG00000205593.12  ENSG00000205593.12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "def bar():\n",
    "    print('we can not use the hugo and elife genes list they are not order correctly')\n",
    "    hg = HUGO_Genes[0:5]\n",
    "    print( hg )\n",
    "    eg = elifeGenes[0:5]\n",
    "    print(eg )\n",
    "\n",
    "    selectRows = mapDF.loc[:,'HUGO_v35'].isin( hg )\n",
    "    display( mapDF.loc[selectRows, :] )\n",
    "   \n",
    "\n",
    "bar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "543778d1-2e14-40be-91a6-7ec79985f860",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(orderedFeatureHugoList) : 712\n",
      "featuredHugoMappingDF.shape : (715, 3)\n",
      "\n",
      "duplicate HUGO_v35 rows\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HUGO_v35</th>\n",
       "      <th>ENSG_v35</th>\n",
       "      <th>ENSG_v39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>670</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.14</td>\n",
       "      <td>ENSG00000182378.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.14</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>693</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.14_PAR_Y</td>\n",
       "      <td>ENSG00000182378.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>694</th>\n",
       "      <td>PLCXD1</td>\n",
       "      <td>ENSG00000182378.14_PAR_Y</td>\n",
       "      <td>ENSG00000182378.15_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    HUGO_v35                  ENSG_v35                  ENSG_v39\n",
       "670   PLCXD1        ENSG00000182378.14        ENSG00000182378.15\n",
       "671   PLCXD1        ENSG00000182378.14  ENSG00000182378.15_PAR_Y\n",
       "693   PLCXD1  ENSG00000182378.14_PAR_Y        ENSG00000182378.15\n",
       "694   PLCXD1  ENSG00000182378.14_PAR_Y  ENSG00000182378.15_PAR_Y"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uniquFeaturedHugoMappingDF.shape : (712, 3)\n",
      "orderedMappingDF.shape : (712, 2)\n",
      "head\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG_v35</th>\n",
       "      <th>ENSG_v39</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUGO_v35</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>(GGTG)n</th>\n",
       "      <td>(GGTG)n</td>\n",
       "      <td>(GGTG)n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(GT)n</th>\n",
       "      <td>(GT)n</td>\n",
       "      <td>(GT)n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TA)n</th>\n",
       "      <td>(TA)n</td>\n",
       "      <td>(TA)n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TCCAC)n</th>\n",
       "      <td>(TCCAC)n</td>\n",
       "      <td>(TCCAC)n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>(TCTATG)n</th>\n",
       "      <td>(TCTATG)n</td>\n",
       "      <td>(TCTATG)n</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG_v35   ENSG_v39\n",
       "HUGO_v35                       \n",
       "(GGTG)n      (GGTG)n    (GGTG)n\n",
       "(GT)n          (GT)n      (GT)n\n",
       "(TA)n          (TA)n      (TA)n\n",
       "(TCCAC)n    (TCCAC)n   (TCCAC)n\n",
       "(TCTATG)n  (TCTATG)n  (TCTATG)n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tail\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG_v35</th>\n",
       "      <th>ENSG_v39</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUGO_v35</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZNF781</th>\n",
       "      <td>ENSG00000196381.11</td>\n",
       "      <td>ENSG00000196381.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNRF1</th>\n",
       "      <td>ENSG00000186187.12</td>\n",
       "      <td>ENSG00000186187.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZRANB1</th>\n",
       "      <td>ENSG00000019995.6</td>\n",
       "      <td>ENSG00000019995.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSWIM4</th>\n",
       "      <td>ENSG00000132003.9</td>\n",
       "      <td>ENSG00000132003.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "      <td>ENSG00000162378.13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSG_v35            ENSG_v39\n",
       "HUGO_v35                                        \n",
       "ZNF781    ENSG00000196381.11  ENSG00000196381.11\n",
       "ZNRF1     ENSG00000186187.12  ENSG00000186187.12\n",
       "ZRANB1     ENSG00000019995.6   ENSG00000019995.6\n",
       "ZSWIM4     ENSG00000132003.9  ENSG00000132003.10\n",
       "ZYG11B    ENSG00000162378.13  ENSG00000162378.13"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "V39FeatureOrderSeries.shape : (712,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "HUGO_v35\n",
       "(GGTG)n        (GGTG)n\n",
       "(GT)n            (GT)n\n",
       "(TA)n            (TA)n\n",
       "(TCCAC)n      (TCCAC)n\n",
       "(TCTATG)n    (TCTATG)n\n",
       "Name: ENSG_v39, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "HUGO_v35\n",
       "ZNF781    ENSG00000196381.11\n",
       "ZNRF1     ENSG00000186187.12\n",
       "ZRANB1     ENSG00000019995.6\n",
       "ZSWIM4    ENSG00000132003.10\n",
       "ZYG11B    ENSG00000162378.13\n",
       "Name: ENSG_v39, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def getV39FeatureOrder():\n",
    "    # select the orderedFeatureHugoList genes from teh mapDF\n",
    "    selectRows = mapDF.loc[:, 'HUGO_v35'].isin( orderedFeatureHugoList )\n",
    "    featuredHugoMappingDF = mapDF.loc[selectRows, :]\n",
    "\n",
    "    # are the HUGO value unique ? NO\n",
    "    print(f'len(orderedFeatureHugoList) : {len(orderedFeatureHugoList)}')\n",
    "    print(f'featuredHugoMappingDF.shape : {featuredHugoMappingDF.shape}')\n",
    "\n",
    "    # https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html\n",
    "    # keep=False will cause a boolean series to be return, where duplicates are marked as true\n",
    "    selectDuplicateRows = featuredHugoMappingDF.duplicated( ['HUGO_v35'], keep=False ) \n",
    "    print(f'\\nduplicate HUGO_v35 rows')\n",
    "    display( featuredHugoMappingDF.loc[selectDuplicateRows, :] )\n",
    "\n",
    "    # \n",
    "    # selectFirstRows = featuredHugoMappingDF.duplicated( ['HUGO_v35'], keep='first' ) \n",
    "    # print( sum(selectFirstRows) )\n",
    "\n",
    "    uniquFeaturedHugoMappingDF = featuredHugoMappingDF.drop_duplicates(subset=['HUGO_v35'], inplace=False)\n",
    "    print(f'uniquFeaturedHugoMappingDF.shape : {uniquFeaturedHugoMappingDF.shape}')\n",
    "\n",
    "    # set the index so we can reorder rows\n",
    "    tmpMappingDF = uniquFeaturedHugoMappingDF.set_index('HUGO_v35', inplace=False)\n",
    "\n",
    "    orderedMappingDF = tmpMappingDF.loc[orderedFeatureHugoList, :]\n",
    "    print(f'orderedMappingDF.shape : {orderedMappingDF.shape}')\n",
    "\n",
    "    print(\"head\")\n",
    "    display( orderedMappingDF.head() )\n",
    "    print(\"tail\")\n",
    "    display( orderedMappingDF.tail() )\n",
    "\n",
    "    retSeries = orderedMappingDF.loc[:, \"ENSG_v39\"]\n",
    "\n",
    "    return retSeries\n",
    "\n",
    "V39FeatureOrderSeries = getV39FeatureOrder()\n",
    "print(f'V39FeatureOrderSeries.shape : {V39FeatureOrderSeries.shape}')\n",
    "display( V39FeatureOrderSeries[0:5] )\n",
    "display( V39FeatureOrderSeries[-5:] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e6c6f27f-2d07-4c98-9ebb-76d7b781e129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XXXDF.shape : (224, 712)\n"
     ]
    }
   ],
   "source": [
    "XXXDF = tmpXDF.loc[:, V39FeatureOrderSeries.values]\n",
    "# model expects the names to be in Hugo Format\n",
    "XXXDF.columns = orderedFeatureHugoList\n",
    "print(f'XXXDF.shape : {XXXDF.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b73610f-fe64-42df-9531-dc0a250d1636",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f808774b-f802-42a7-a4e7-c10bf9722f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# V39FeatureOrderSeries = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5b1a5fc7-0e49-44b3-8089-e200fa9b773d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# display( XDF.iloc[:, 0:5].head() )\n",
    "# t = V39FeatureOrderSeries[0:5]\n",
    "# print()\n",
    "# display( XDF.loc[:, t].head() )\n",
    "\n",
    "# XXXDF = XDF.loc[:, V39FeatureOrderSeries]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2c402e66-e685-41bb-abb7-35706d8ff5e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #V39FeatureOrderSeries['ENSG00000182378.15_PAR_Y']\n",
    "# #         \"ENSG00000182378.15_PAR_Y\" : \"ENSG00000182378.15\",\n",
    "\n",
    "# def foo():\n",
    "#     # \tHUGO_v35 ENSG_v35\tENSG_v39\n",
    "#     #select = mapDF.loc[:, 'HUGO_v35'] == \"ENSG00000182378.15\"\n",
    "#     #select = mapDF.loc[:, 'ENSG_v35'] == \"ENSG00000182378.15\"\n",
    "#     select = mapDF.loc[:, 'ENSG_v39'] == \"ENSG00000182378.15\"\n",
    "#     display( mapDF.loc[select, :] )\n",
    "\n",
    "# foo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "246f0d84-0cad-4be7-b4fd-dcc942e590c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "39d7b669-fb71-4ada-a98c-890e44f0e118",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "predictions:\n",
      "[82 82 82 82 82]\n",
      "\n",
      "yProbability:\n",
      "[[0.05 0.   0.   0.   0.01 0.01 0.02 0.01 0.   0.   0.   0.   0.01 0.01\n",
      "  0.   0.   0.   0.   0.   0.04 0.01 0.01 0.   0.   0.   0.   0.   0.\n",
      "  0.02 0.   0.   0.   0.01 0.   0.   0.   0.   0.   0.01 0.   0.1  0.\n",
      "  0.   0.01 0.03 0.   0.01 0.01 0.   0.01 0.06 0.   0.   0.05 0.   0.\n",
      "  0.   0.   0.   0.   0.01 0.   0.   0.   0.   0.   0.   0.   0.   0.\n",
      "  0.   0.   0.   0.01 0.   0.   0.02 0.   0.   0.   0.   0.   0.46]\n",
      " [0.02 0.   0.   0.   0.02 0.   0.02 0.   0.   0.   0.01 0.   0.05 0.01\n",
      "  0.   0.   0.   0.01 0.01 0.01 0.03 0.   0.   0.01 0.   0.   0.01 0.02\n",
      "  0.02 0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.01 0.1  0.\n",
      "  0.   0.   0.   0.   0.02 0.01 0.   0.05 0.   0.   0.   0.05 0.   0.\n",
      "  0.   0.   0.01 0.   0.04 0.   0.   0.01 0.   0.   0.   0.02 0.   0.\n",
      "  0.   0.   0.   0.   0.   0.   0.01 0.   0.   0.02 0.   0.   0.4 ]\n",
      " [0.05 0.01 0.   0.   0.   0.   0.01 0.   0.   0.   0.   0.   0.04 0.\n",
      "  0.   0.   0.   0.   0.   0.   0.01 0.   0.01 0.   0.   0.   0.01 0.\n",
      "  0.04 0.   0.   0.   0.01 0.   0.   0.   0.   0.01 0.   0.   0.23 0.\n",
      "  0.01 0.   0.   0.   0.02 0.   0.   0.02 0.03 0.   0.   0.03 0.   0.\n",
      "  0.   0.   0.   0.   0.05 0.   0.   0.01 0.   0.   0.   0.02 0.   0.\n",
      "  0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.38]\n",
      " [0.03 0.   0.   0.   0.   0.   0.02 0.03 0.   0.   0.01 0.   0.05 0.\n",
      "  0.   0.   0.   0.   0.02 0.02 0.   0.01 0.   0.   0.   0.   0.   0.01\n",
      "  0.04 0.   0.   0.   0.   0.   0.   0.03 0.   0.   0.   0.   0.22 0.\n",
      "  0.02 0.   0.   0.   0.02 0.   0.   0.02 0.02 0.   0.   0.01 0.   0.\n",
      "  0.   0.   0.   0.01 0.06 0.   0.   0.01 0.   0.   0.   0.   0.   0.\n",
      "  0.   0.   0.   0.01 0.   0.   0.01 0.   0.   0.03 0.   0.   0.29]\n",
      " [0.02 0.   0.   0.02 0.   0.01 0.   0.01 0.   0.   0.02 0.   0.02 0.\n",
      "  0.   0.   0.   0.   0.01 0.01 0.   0.04 0.   0.   0.   0.   0.   0.01\n",
      "  0.05 0.   0.   0.   0.02 0.   0.   0.03 0.   0.   0.   0.02 0.1  0.\n",
      "  0.   0.   0.   0.   0.03 0.01 0.   0.01 0.05 0.   0.01 0.05 0.01 0.\n",
      "  0.   0.   0.   0.   0.06 0.01 0.   0.   0.   0.01 0.   0.02 0.   0.\n",
      "  0.   0.02 0.   0.   0.   0.   0.   0.   0.   0.01 0.   0.   0.31]]\n",
      "CPU times: user 60.9 ms, sys: 144 µs, total: 61.1 ms\n",
      "Wall time: 60.1 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "predictions  = model.predict(XXXDF)\n",
    "print(f'\\npredictions:\\n{predictions[0:5]}')\n",
    "\n",
    "yProbability = model.predict_proba(XXXDF)\n",
    "print(f'\\nyProbability:\\n{yProbability[0:5]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e52575e3-4a8a-4aaf-9810-79552ea0c4f7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "44cc250f-2725-4d1e-98d1-7f5a0238b4e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rename labelEncoder_Junk to labelEncoderElife\n",
      " number of Whole_Blood predictions : 205  92%\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>elife</th>\n",
       "      <th>prediction</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Colorectal Cancer</th>\n",
       "      <th>Heart_Left_Ventricle</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Esophagus Cancer</th>\n",
       "      <th>Heart_Left_Ventricle</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Healthy donor</th>\n",
       "      <th>Heart_Left_Ventricle</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Liver Cancer</th>\n",
       "      <th>Heart_Left_Ventricle</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Testis</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Lung Cancer</th>\n",
       "      <th>Testis</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Stomach Cancer</th>\n",
       "      <th>Heart_Left_Ventricle</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Whole_Blood</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         i\n",
       "elife             prediction              \n",
       "Colorectal Cancer Heart_Left_Ventricle   4\n",
       "                  Whole_Blood           49\n",
       "Esophagus Cancer  Heart_Left_Ventricle   2\n",
       "                  Whole_Blood           29\n",
       "Healthy donor     Heart_Left_Ventricle   3\n",
       "                  Whole_Blood           40\n",
       "Liver Cancer      Heart_Left_Ventricle   2\n",
       "                  Testis                 5\n",
       "                  Whole_Blood           19\n",
       "Lung Cancer       Testis                 2\n",
       "                  Whole_Blood           33\n",
       "Stomach Cancer    Heart_Left_Ventricle   1\n",
       "                  Whole_Blood           35"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>i</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prediction</th>\n",
       "      <th>elife</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Heart_Left_Ventricle</th>\n",
       "      <th>Colorectal Cancer</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Esophagus Cancer</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy donor</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Liver Cancer</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stomach Cancer</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Testis</th>\n",
       "      <th>Liver Cancer</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lung Cancer</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Whole_Blood</th>\n",
       "      <th>Colorectal Cancer</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Esophagus Cancer</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Healthy donor</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Liver Cancer</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lung Cancer</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stomach Cancer</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         i\n",
       "prediction           elife                \n",
       "Heart_Left_Ventricle Colorectal Cancer   4\n",
       "                     Esophagus Cancer    2\n",
       "                     Healthy donor       3\n",
       "                     Liver Cancer        2\n",
       "                     Stomach Cancer      1\n",
       "Testis               Liver Cancer        5\n",
       "                     Lung Cancer         2\n",
       "Whole_Blood          Colorectal Cancer  49\n",
       "                     Esophagus Cancer   29\n",
       "                     Healthy donor      40\n",
       "                     Liver Cancer       19\n",
       "                     Lung Cancer        33\n",
       "                     Stomach Cancer     35"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# evaluate predictions\n",
    "elifeLabels = labelEncoderElife.inverse_transform(yNP)\n",
    "predictionLabels = labelEncoder.inverse_transform( predictions )\n",
    "\n",
    "numWholeBlood = sum(predictionLabels == \"Whole_Blood\")\n",
    "print(f' number of Whole_Blood predictions : {numWholeBlood}  {round( 100 * numWholeBlood / len(predictionLabels))}%' )\n",
    "\n",
    "\n",
    "predictionDF = pd.DataFrame( {'elife' : elifeLabels, \n",
    "                               'prediction' : predictionLabels,\n",
    "                                'i' : [1 for i in range(len(yNP))] } )\n",
    "\n",
    "\n",
    "display( predictionDF.groupby( [\"elife\", 'prediction'] ).count() )\n",
    "\n",
    "print()\n",
    "display( predictionDF.groupby( ['prediction', \"elife\"] ).count() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f940185-7d7e-458a-90e6-ef51765b701c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931bddff-ceb3-41c4-8078-f72456039772",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8524fcce-bad5-490c-a972-f42cc445fa5a",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (2149445957.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[21], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    aedwip do not plot confusion matrix. it just a big mess\u001b[0m\n\u001b[0m           ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "aedwip do not plot confusion matrix. it just a big mess"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae45088-3ffb-49ea-a987-81dce6019489",
   "metadata": {},
   "source": [
    "## extraCellularRNA/terra/jupyterNotebooks/cibersort/fractionsAsMulticlassClassification.ipynb\n",
    "copied ploting code from advancement. make_confusion_matrix() did not work well when we have a large # of classes\n",
    "\n",
    "**<span style=\"color:red;background-color:yellow\">refactor notebooks</span>**  \n",
    "\n",
    "- deconvolutionAnalysis/jupyterNotebooks/randomForestGeneSignatureDeconvolution.ipynb\n",
    "- intraExtraRNA_POC/jupyterNotebooks/elife/randomForestGTEx_TCGAGeneSignature.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4def5a18-1e40-42cd-82e5-102b9d0a3161",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from confusion_matrix.cf_matrix import make_confusion_matrix\n",
    "\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "from pipeline.dataFactory.utilities import urlify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d2e2888-c3b7-4964-b933-0fdb4d3b9d5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bme263DataVis.utilities as utl\n",
    "import seaborn as sns\n",
    "\n",
    "def plotConfusionMatrix(confusionData, labels, annotations, imgOutDir, title, pageWidthInInches, \n",
    "                        pageHeightInInches, displayCounts=False):\n",
    "    '''\n",
    "    arguments\n",
    "        confusionData, labels, \n",
    "        \n",
    "        annotations\n",
    "            string values to display in heat map cells, must have same shape as confusionData\n",
    "        \n",
    "        imgOutDir, title, pageWidthInInches, pageHeightInInches\n",
    "        \n",
    "        displayCounts:\n",
    "            boolean: default = False\n",
    "            if true fmt is set to '.0f', and annotations are not displayed\n",
    "    '''\n",
    "    \n",
    "    plotUtils =  utl.MatPlotLibUtilities()\n",
    "    fig = plt.figure( figsize=(pageWidthInInches, pageHeightInInches) )\n",
    "\n",
    "    # correlation\n",
    "    cbarPanelWidthInInches = 0.25 #0.5\n",
    "    panelWidthInInches = pageWidthInInches - cbarPanelWidthInInches\n",
    "    panelHeightInInches = pageHeightInInches \n",
    "    leftRelativeSize = 0\n",
    "    bottomRelativeSize = 0\n",
    "    panel = plotUtils.createPanel(fig, panelWidthInInches, panelHeightInInches, \n",
    "                                             leftRelativeSize, bottomRelativeSize)\n",
    "    \n",
    "    cbarPanelHeightInInches = panelHeightInInches - 2 #1\n",
    "    cbarLeftRelativeSize = 0.99 #0.985 #0.975 # 0.95   # % of fig width\n",
    "    cbarBottomRelativeSize = 0.08 #0.1 #0.125 #0.15 # 0.10 # % of fig height\n",
    "    cbarPanel = plotUtils.createPanel(fig, cbarPanelWidthInInches, cbarPanelHeightInInches, \n",
    "                                             cbarLeftRelativeSize, cbarBottomRelativeSize)\n",
    "\n",
    "    if (displayCounts):\n",
    "        panel = sns.heatmap( confusionData \n",
    "                            , xticklabels = labels\n",
    "                            , yticklabels = labels\n",
    "                            , ax = panel\n",
    "                            , cbar_ax = cbarPanel\n",
    "                            , cmap = \"Blues\"\n",
    "                            #, cmap=\"viridis\"\n",
    "                            , annot = True # display the numeric values in confusionData\n",
    "                            #, fmt='.2%' # use if annot = True # default is '.2g'\n",
    "                            , fmt=\".0f\" # use to display true positive counts\n",
    "                            , square=True\n",
    "                            , linecolor = \"black\" # default is \"white\"\n",
    "                            , linewidths = 1.0\n",
    "                   )\n",
    "    else:\n",
    "        panel = sns.heatmap( confusionData \n",
    "                        , xticklabels = labels\n",
    "                        , yticklabels = labels\n",
    "                        , ax = panel\n",
    "                        , cbar_ax = cbarPanel\n",
    "                        , cmap = \"Blues\"\n",
    "                        #, cmap=\"viridis\"                            \n",
    "                        , annot = annotations\n",
    "                        , fmt='' # required if our annotations are strings\n",
    "                        , square=True\n",
    "                        , linecolor = \"black\" # default is \"white\"\n",
    "                        , linewidths = 1.0\n",
    "\n",
    "               )        \n",
    "\n",
    "    panel.set_title(title)\n",
    "    panel.set_ylabel('Actual')\n",
    "    panel.set_xlabel('Predicted')\n",
    "\n",
    "    # https://drawingfromdata.com/seaborn/matplotlib/visualization/rotate-axis-labels-matplotlib-seaborn.html\n",
    "    panel.set_xticklabels(panel.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "\n",
    "    # imgFile = imgOutDir.joinpath( urlify(title) + \".png\" )\n",
    "    imgFile = f'{imgOutDir}/{urlify(title) + \".png\"}'\n",
    "    print(f\"save: {imgFile}\")\n",
    "    plt.savefig( imgFile , bbox_inches='tight', dpi=300,  facecolor=\"white\", transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4c6f793-e3ff-4817-9c0c-0f489afa0d89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotConfusionMatrixHelper(imgOutDir, title, cfMatrix, labels, start, end, displayCounts=False):\n",
    "    pageWidthInInches  = 8\n",
    "    pageHeightInInches = 8\n",
    "    \n",
    "    if displayCounts:\n",
    "        confusionData = cfMatrix\n",
    "        annot = None\n",
    "    else :\n",
    "        confusionData = calculateRowPercentages( cfMatrix )\n",
    "        annotations = createHeatMapAnnotations( confusionData )\n",
    "        annot = annotations[start:end, start:end]\n",
    "\n",
    "    cfMatrixL = confusionData[start:end, start:end] # confusion_matrix is TP, FP counts we want percentages \n",
    "    labelsL = labels[start:end]\n",
    "    plotConfusionMatrix(cfMatrixL, labelsL, annot, imgOutDir, title, pageWidthInInches, \n",
    "                        pageHeightInInches, displayCounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd167386-ebff-4e4b-a318-23028676a1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculateRowPercentages(cm):\n",
    "    '''\n",
    "    divides each value in a row by the row totoal\n",
    "    \n",
    "    arguments:\n",
    "        cm: a confusion matrix\n",
    "        \n",
    "    returns\n",
    "        a numpy array with same shape as cm \n",
    "    '''\n",
    "    byRows = 1\n",
    "    rowMarginEquals1 = np.sum(cm, axis=byRows)\n",
    "\n",
    "    rowMarginEquals1 = rowMarginEquals1.astype(np.float16)\n",
    "    \n",
    "    print(f'AEDWIP hack')\n",
    "    for i in range(len(rowMarginEquals1)):\n",
    "        v = rowMarginEquals1[i]\n",
    "        print(f' i: {i} v:{v}')\n",
    "        if v == 0:\n",
    "            rowMarginEquals1[i] = 999\n",
    "            print(f'aeDWIP hack i: {i} v: {v} rowMarginEquals1[i] : {rowMarginEquals1[i]} ' )\n",
    "    \n",
    "    print(f\"aediwp rowMarginEquals1\\n{rowMarginEquals1}\")\n",
    "    print(f\"aedwip rowMarginEquals1.shape\\n{rowMarginEquals1.shape}\")\n",
    "\n",
    "    rowMarginEquals1Reshape = rowMarginEquals1.reshape(rowMarginEquals1.shape[0], 1)\n",
    "    #print(f\"\\n rowMarginEquals1Reshape\\n{rowMarginEquals1Reshape}\")\n",
    "\n",
    " \n",
    "    print(f'AEDWIP rowMarginEquals1Reshape.shape : {rowMarginEquals1Reshape.shape}')\n",
    "    print(f'AEDWIP cm.shape : {cm.shape}')\n",
    "    print(f'AEDWIP rowMarginEquals1Reshape : \\n{rowMarginEquals1Reshape}')\n",
    "\n",
    "    ret = cm/rowMarginEquals1Reshape\n",
    "    \n",
    "    return ret\n",
    "    \n",
    "    \n",
    "# def testCalculateRowPercentages(cm):\n",
    "#     ret = calculateRowPercentages(cm)\n",
    "#     print(ret)\n",
    "    \n",
    "# testCalculateRowPercentages( cfMatrix )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c337c822-903c-4437-8d8e-8c942ace4233",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createHeatMapAnnotations( rowPercentages ):\n",
    "    '''\n",
    "    seaborn heatmap argument annot = True display the cfMatrix cell values.\n",
    "    fmt='.2%' argument would display the number cell values with 2 decimal places\n",
    "    \n",
    "    our confustion matrix is large, and has a lot of zeros making it hard to visualy\n",
    "    interpurt the data.\n",
    "    \n",
    "    retuns \n",
    "        a numpy array of string with the same dimensions as rowPercentages\n",
    "        use this as the value for the argument annot, do not use fmt\n",
    "    '''\n",
    "    numRows = rowPercentages.shape[0]\n",
    "    numCols = rowPercentages.shape[1]\n",
    "    \n",
    "    bufferList = [0] * numRows\n",
    "    for i in range(numRows):        \n",
    "        a = [\"\"] * numCols \n",
    "        for j in range(numCols):\n",
    "            d = rowPercentages[i,j]\n",
    "            if d > 0.0:\n",
    "                #s = '{:.2f}'.format(d)\n",
    "                s = '{}%'.format( round(d * 100), digits=0) \n",
    "                a[j] = s\n",
    "                \n",
    "        aNP = np.array( a, dtype=str )#.reshape( (1,numCols) )\n",
    "        bufferList[i] = aNP\n",
    "        \n",
    "    return np.array(bufferList, dtype=object) \n",
    "    \n",
    "# def testCreateHeatMapAnnotations():\n",
    "#     rowPercentages = calculateRowPercentages(cfMatrix)\n",
    "#     print(f'rowPercentages.shape: {rowPercentages.shape}')\n",
    "#     print(rowPercentages)\n",
    "#     ret = createHeatMapAnnotations( rowPercentages )\n",
    "#     print(f'ret.shape: {ret.shape}')\n",
    "#     print(ret)\n",
    "    \n",
    "# testCreateHeatMapAnnotations()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89ed4edb-779f-418f-948d-4b27010c0378",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f'AEDWIP yNP.shape : {yNP.shape}')\n",
    "print(f'AEDWIP predictions.shape : {predictions.shape}')\n",
    "cm = confusion_matrix(yNP, predictions)\n",
    "confusionData = calculateRowPercentages( cm )\n",
    "labelList = list( labelEncoder.classes_ )\n",
    "labels =  labelList\n",
    "annotations = createHeatMapAnnotations( confusionData )\n",
    "\n",
    "title = f\"Random Forest {modelName} Gene Signature GTEx _TCGA Training Set\"\n",
    "\n",
    "pageWidthInInches = 8\n",
    "pageHeightInInches  = 8\n",
    "plotConfusionMatrix( confusionData, labels, annotations, imgOut, title, pageWidthInInches, \n",
    "                        pageHeightInInches, displayCounts=False )\n",
    "\n",
    "# tmpDir = \"./tmp\"\n",
    "# plotConfusionMatrixHelper(tmpDir, title, cm, labels=['c', 'd',  'e', 'f', 'g', 'a', 'b', 'z'], start=0, end=8, displayCounts=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc5da966-a34c-4c08-b7a7-551a4b1de473",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(cm)\n",
    "\n",
    "print()\n",
    "print(yNP.shape)\n",
    "print(yNP)\n",
    "\n",
    "print()\n",
    "print( labelEncoder_Junk.classes_)\n",
    "\n",
    "print()\n",
    "print(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de38ae0-9d1b-4f46-96b8-0b82653be493",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many samples where predicted to be whole blood\n",
    "sum( predictions == 82 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7380be77-e670-42d8-98ca-58620a6155a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "notWholeBloodNP = predictions[ predictions != 82 ] \n",
    "print(f' {labelEncoder.classes_[40]} ')\n",
    "print(f' {labelEncoder.classes_[75]} ')\n",
    "print(f' {labelEncoder.classes_[82]} ')\n",
    "notWholeBloodNP\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50cc89ea-9a9e-4e0a-bc5a-7c9a00814afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "print()\n",
    "yPDF = pd.DataFrame( yProbability )\n",
    "yPDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d9b595-1928-4065-9670-172f0a53d572",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cfe793b-e645-4f62-8912-d81957d036cf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
