---
title: "Normalzed biotype counts"
author: "Andrew Davidson, aedavids@ucsc.edu"
date: "Oct 14, 2020"

output: html_notebook
---
Uses DESeq package to get comparable transcript counts

overview: 
  - tximport: read estimated salmon quant.sf (transcripts counts)
  - DESeqDataSetFromTximport :
  - estimateSizeFactors :
  - counts()


ref:

- [bioconductor.org vignettes/DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
- [bioconductor BiocWorkshops rna-seq-data-analysis-with-deseq2](https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html)
- /public/home/aedavids/extraCellularRNA/R/notebooks/kras.ipsc_exo_vs_bulk.DESeq.Rmd

- [R class tutorial](https://www.datamentor.io/r-programming/object-class-introduction/)

- /public/groups/kimlab/kras.ipsc/old.output.data.and.scripts/tximport.and.normalize.R


```{r message=FALSE}
library("DESeq2")
library('rjson')
library("tximport")
source("DESeqUtils.R") # assume file in in same directory as current Rmd
source("findFiles.R") # assume file in in same directory as current Rmd
```
## Find all the salmon estimated transcript count files
```{r}
print( sprintf("BEGIN: %s", Sys.time()) )
rootDir <- "/home/kimlab/exoRNA-biomarkers-panc/data/a549_0.2MOI_24hr"
outputRoot <- "/home/aedavids/extraCellularRNA/data/R/output"
dir.create(outputRoot, recursive=TRUE, showWarnings = FALSE)

# ref: salmon cmd_info.json output file captures arguments
# indexFilePath <- file.path('/public/groups/kimlab/indexes/',
#                             'sel.align.gencode.v35.process.aware.salmon.v1.3.0.sidx')

indexFilePath <- file.path( '/home/kimlab/genomes.annotations/gencode.35', 
                            'gencode.v35.ucsc.rmsk.tx.to.gene.csv')
```


# DESeq
We do not actually run "differential expression analysis", we just use some of DESeq function to get
comparable count data

```{r}
########################################################################
findA549_0.2MOI_24hrForCondition <- function(rootDir, condition, index) {
  #
  # finds salmon quant.sf files. Checks to make sure file exists
  #
  # arguments
  #   rootDir: example "/public/groups/kimlab/exoRNA-biomarkers-panc/data/a549_0.2MOI_24hr/"
  #
  #   condition: "SARS" or "Mock"
  #
  #   index: example "sel.align.gencode.v35.salmon.v1.3.0.sidx"
  #
  # returns
  #   a list of full salmon file path
  
  # example file pathys
  # SRR11412240.Mock.6/SRR11412240.Mock.6_sel.align.gencode.v35.salmon.v1.3.0.sidx_2020-10-15-21.32.19-PDT/
  # SRR11412256.SARS.11/SRR11412256.SARS.11_sel.align.gencode.v35.salmon.v1.3.0.sidx_2020-10-15-21.32.19-PDT/

  # find all files 
  
  # TODO try x.path <- Sys.glob(file.path(rootDir, "*process.aware.salmon*", "quant.sf"))[1]
  lf <- list.files(rootDir, recursive=TRUE, full.names=TRUE)
  
  # grep returns index, llf wil be a files paths containing condition
  llf <- lf[ grep(condition, lf)]
  lllf <- llf[ grep(index, llf)]
  salmonFiles <- lllf[grep("quant.sf", lllf)]
  
  
  if ( ! all(file.exists(salmonFiles)) ) {
    stop(c("ERROR: file not found: ", salmonFiles))
  }

  return(salmonFiles)
}

findA549_0.2MOI_24hrFiles <- function(rootDir) {
  dataFilesList = list()
  
  # TODO figure out how to pass args to getNormalizedDESeqDataSet()
  index = "sel.align.gencode.v35.salmon.v1.3.0.sidx"
  
  mockSalmonFiles <- findA549_0.2MOI_24hrForCondition(rootDir, "Mock", index)
  dataFilesList$mockSalmonFiles <- mockSalmonFiles
  
  SARSSalmonFiles <- findA549_0.2MOI_24hrForCondition(rootDir, "SARS", index)
  dataFilesList$SARSSalmonFiles <- SARSSalmonFiles
  
  return( dataFilesList)
}
```


```{r}
getA549_0.2MOI_24hrColDataDF <- function (fileList) {
  # create the colData data frame. This is the meta data describing each sample
  # The Rows of correspond to columns of countData and cols are features that
  # describe the replicates. The design model use the columns as features
  
  numMock <- length( fileList$mockSalmonFiles )
  numSARS  <- length( fileList$SARSSalmonFiles )
  
  diseaseState <- c( rep('Mock', numMock ),
                     rep('SARS', numSARS )) 
  
  diseaseState <- factor(diseaseState, levels=c("Mock", "SARS"))
  
  colDF <- data.frame(diseaseState)
  
  return(colDF)  
}
```


Construct a DESeq data object for biotypes
```{r}
  
#index = "sel.align.gencode.v35.salmon.v1.3.0.sidx"
## allSalmonQuantFiles <- unlist(A549_0.2MOI_24hrFilesList, use.names = FALSE)
# 
# colDF <- getA549_0.2MOI_24hrColDataDF( A549_0.2MOI_24hrFilesList )


mapToId <- SALMON_QUANT_SF_NAMES_LIST()$BIO_TYPE
mapToIdName <- names(SALMON_QUANT_SF_NAMES_LIST()[mapToId])
# bioTypeDDS <-
#   getNormalizedDESeqDataSet(rootDir,
#                           indexFilePath,
#                           allSalmonQuantFiles, 
#                           colDF,
#                           mapToId,
#                           design=~ diseaseState)

# warning we are not using the model. 
retList <-
  getNormalizedDESeqDataSet(rootDir,
                          indexFilePath,
                          findFilesFunc=findA549_0.2MOI_24hrFiles, 
                          getColDFFunc=getA549_0.2MOI_24hrColDataDF,
                          design=~ diseaseState,
                          mapToId
                          )

bioTypeDDS <- retList$dds
fileList <- retList$fileList

#csvOutFileName <- "A549_0.2MOI_24hr.normalized.deseq.biotype.counts.csv"
csvBioTypeOutFileName <- paste0("A549_0.2MOI_24hr.normalized.deseq.", mapToIdName, ".counts.csv")

#aediwp 5,4 was when we had 6 replicants in 2 conditions
bioTypeDF <- createComparableCounts( bioTypeDDS, filterRowCount=5, filterSampleCount=4 )
```

construct DESeq data object for Ensembl transcript stable id  
```{r}
mapToId <- SALMON_QUANT_SF_NAMES_LIST()$ENST
mapToIdName <- names(SALMON_QUANT_SF_NAMES_LIST()[mapToId])

retList <-
  getNormalizedDESeqDataSet(rootDir,
                          indexFilePath,
                          findFilesFunc=findA549_0.2MOI_24hrFiles, 
                          getColDFFunc=getA549_0.2MOI_24hrColDataDF,
                          design=~ diseaseState,
                          mapToId
                          )

enstDDS <- retList$dds
fileList <- retList$fileList

#csvOutFileName <- "A549_0.2MOI_24hr.normalized.deseq.biotype.counts.csv"
csvENSTOutFileName <- paste0("A549_0.2MOI_24hr.normalized.deseq.", mapToIdName, ".counts.csv")

#aediwp 5,4 was when we had 6 replicants in 2 conditions
enstDDSDF <- createComparableCounts( enstDDS, filterRowCount=5, filterSampleCount=4 )
```


```{r}
getCountDataColNames <- function( fileList ) {
  #
  # arguments
  #   fileList
  #     example of an element
  #     "/home/kimlab/pancreas.plasma.ev.long.RNA/data/PDAC/SRR9624965/salmon.out/quant.sf"
  #
  # returns a list of colData column names. format diseaseState_sampleIs
  #   example
  #     "healthy_SRR10080507" "healthy_SRR10080508" "healthy_SRR10080509" 

  tokensListList <- strsplit( fileList, split="/" )
  # [[1]]
  # [1] ""                                                                                   
  # [2] "home"                                                                               
  # [3] "kimlab"                                                                             
  # [4] "exoRNA-biomarkers-panc"                                                             
  # [5] "data"                                                                               
  # [6] "a549_0.2MOI_24hr"                                                                   
  # [7] "SRR11412239.Mock.2"                                                                 
  # [8] "SRR11412239.Mock.2_sel.align.gencode.v35.salmon.v1.3.0.sidx_2020-10-15-21.32.19-PDT"
  # [9] "quant.sf"                                                                           

      
  
  tokensList <- unlist(tokensListList)
  # [1] ""                                                                                    
  # [2] "home"                                                                                
  # [3] "kimlab"                                                                              
  # [4] "exoRNA-biomarkers-panc"                                                              
  # [5] "data"                                                                                
  # [6] "a549_0.2MOI_24hr"                                                                    
  # [7] "SRR11412239.Mock.2"                                                                  
  # [8] "SRR11412239.Mock.2_sel.align.gencode.v35.salmon.v1.3.0.sidx_2020-10-15-21.32.19-PDT" 
  # [9] "quant.sf"                                                                            
  # [10] ""              
  
  m <- matrix( tokensList, ncol=9, byrow=TRUE)
  
  diseaseStateIdx <- 6
  sampleIdIdx <-7
  metaTokensMatrix <- m[, c(sampleIdIdx)]
  sampleTokens <- strsplit(metaTokensMatrix, split="\\.")
  m <- matrix( unlist(sampleTokens), ncol=3, byrow=TRUE)
  
  retList <- paste(m[,2], m[,1], sep='_')
  # [1] "Mock_SRR11412239" "Mock_SRR11412240" "Mock_SRR11412241" 

  return( retList )
}
```

change column name to format "diseaseState_sampleId". if we have a
small number of rows, ie. biotype, it would be easier to work with the data if
it was denormalized. that is to say keep sampleType, treatment, and replicant id
as separate columns. If we are in gene space this would add a lot of redundant data
```{r}

metaColNames1 <- getCountDataColNames(fileList$mockSalmonFiles)
metaColNames2 <- getCountDataColNames(fileList$SARSSalmonFiles)
colnames(bioTypeDF) <- c(metaColNames1, metaColNames2)
colnames(enstDDSDF) <- c(metaColNames1, metaColNames2)

```

```{r}
#head( df )

# save
#outFiles <- c(csvBioTypeOutFileName, csvENSTOutFileName)
outFiles <- list( list(bioTypeDF, csvBioTypeOutFileName) , list(enstDDSDF, csvENSTOutFileName) )
for (outfileItem in outFiles) {
  df <- outfileItem[1]
  outfile <- outfileItem[2]
  csvFilePath <- file.path( outputRoot, outfile)
  print(csvFilePath)
  write.csv(df, csvFilePath)
}

print( sprintf("END: %s", Sys.time()) )
```

