---
title: "pancreas.plasma.ev.long.RNA Differential Gene Expression"
author: "Andrew Davidson, aedavids@ucsc.edu"
date: "Dec 28, 2020"

output: html_notebook
---

```{r message=FALSE}
startTime <- Sys.time()
print( sprintf("BEGIN: %s", Sys.time()) )

library("DESeq2")
# library("readr")
# library("tximport")
source("DESeqUtils.R") # assume file in in same directory as current Rmd
source("findFiles.R") # assume file in in same directory as current Rmd
source("te.R") # assume file in in same directory as current Rmd
```



```{r}
debug.parameters <- list(
  'rootDir' = "/home/aedavids/extraCellularRNA/data/debugTximport/pancreas.plasma.ev.long.RNA",
  'findFilesFunc' = findPancreasPlasmaEV_longRNAFiles, 
  'getColDFFunc' = getPancreasPlasmaEV_longRNAColDataDF,    
  'tx2MappingFile' = 'debugTx2Gene.csv',
  'tx2MappingDir' = file.path('/home/aedavids/extraCellularRNA/data/debugTximport',
                         'genomes.annotations/gencode.v35'),  
  # 'outputRoot' = "/home/aedavids/extraCellularRNA/data/R/output/debug",
  'outputRoot' = "/home/aedavids/extraCellularRNA/data/R/debugTximport/output",
  'mapToIdName'='gene',
  'transcriptMapperDFFunc'=getTEGeneMapper,    
  
  # ignoreAfterBar logical, whether to split the tx id on the '|' character to 
  # facilitate matching with the tx id in tx2gene (default FALSE)
  #'ignoreAfterBar'=FALSE #bug we do not get gene names in output
  'ignoreAfterBar'=FALSE,
  
  #  txOut     avoid gene-level summarization by setting txOut=TRUE
  #       if you are trying to normalize transcript and not map them to
  #       genes, or biotype, setting txOut=TRUE reduces required memory
  #       this allows use gencode.v35.ucsc.rmsk.tx.to.gene.csv    
  'txOut'=FALSE
)

TE.gene.parameters <- list(
    'rootDir' = "/home/kimlab/pancreas.plasma.ev.long.RNA",
    'findFilesFunc' = findPancreasPlasmaEV_longRNAFiles, 
    'getColDFFunc' = getPancreasPlasmaEV_longRNAColDataDF,        
    'tx2MappingFile'='gencode.v35.ucsc.rmsk.tx.to.gene.csv',
    'tx2MappingDir' = file.path('/home/kimlab/genomes.annotations/gencode.35'),
    'mapToIdName'='gene',
    'outputRoot' = "/home/aedavids/extraCellularRNA/data/R/output",
    #'outputRoot' = "/home/kimlab/aedavids/extraCellularRNA/data/R/output",
    'transcriptMapperDFFunc'=getTEGeneMapper,
    # ignoreAfterBar logical, whether to split the tx id on the '|' character to 
    # facilitate matching with the tx id in tx2gene (default FALSE)
    #'ignoreAfterBar'=FALSE #bug we do not get gene names in output
    'ignoreAfterBar'=FALSE,
    'txOut'=FALSE
    )

# select parameters to run
#parameters <- TE.gene.parameters
parameters <- debug.parameters
```

set up output locations
```{r}
# dateSetName Used in plot titles
dataSetName <- "pancreas.plasma.ev.long.RNA"

imgDir <- file.path(parameters$outputRoot, parameters$tx2MappingFile, dataSetName, "img")
dir.create(imgDir, recursive=TRUE, showWarnings = FALSE)
cat(sprintf("images will be saved to \n%s",imgDir))

dataOutDir <- file.path(parameters$outputRoot, parameters$tx2MappingFile, dataSetName)
dir.create(dataOutDir, recursive=TRUE, showWarnings = FALSE)
cat(sprintf("LCF data will be saved to \n%s", dataOutDir))
```

## Load data frame that maps from transcript id to gene name
```{r}
getTranscriptMapper <- function(parameters, debug=FALSE) {
  tx2MappingFile <- parameters$tx2MappingFile
  tx2MappingDir <- parameters$tx2MappingDir
  transcriptMapperDFFunc <- parameters$transcriptMapperDFFunc
  
  if (debug) {
    print(sprintf("tx2MappingFile: %s", tx2MappingFile))
    print(sprintf("tx2MappingDir: %s", tx2MappingDir))
  }
  
  transcriptMapperDF <- transcriptMapperDFFunc(tx2MappingDir, tx2MappingFile)
  
  return( transcriptMapperDF )
}
  
transcriptMapperDF <- getTranscriptMapper( parameters, debug=TRUE)
```

## import salmon quant files and create DESeqDataSet
```{r}
# https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
# In order to benefit from the default settings of the package, you should put
# the variable of interest at the end of the formula and make sure the control
# level is the first level.
design <- ~ diseaseState

AEDWIP_indexFilePath_deprecated <- NULL # TODO it is no longer used 
ddsList <- getDESeqDataSet(rootDir=parameters$rootDir,
                           tx2MappingFilePath,
                          findFilesFunc= parameters$findFilesFunc, 
                          getColDFFunc=parameters$getColDFFunc,
                          design = design,
                          transcriptMapperDF,
                          ignoreAfterBar=parameters$ignoreAfterBar,
                          txOut=parameters$txOut)
  
dds <- ddsList$dds
fileList <- ddsList$fileList
```


## pre filtering:
ref: https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering
pre-filter low count genes is not necessary.
the reasons which make pre-filtering useful:
  1) reduce the memory of dds data object
  2) increase the speed of the transformation and testing functions within DESeq2

https://support.bioconductor.org/p/65256/#65260
"you typically don't need to pre-filter because independent filtering occurs
within results() to save you from multiple test correction on genes with no
power (see ?results and the vignette section about independent filtering, or
the paper). The main reason to pre-filter would be to increase speed.
Designs with many samples and many interaction terms can slow down on genes
which have very few reads.

The reason results() filters is having fewer genes improves the adjusted p-values.


## Run DESeq() & Extract the resuls
```{r}
dds <- DESeq(dds)
```

results() extracts a results table with log2 fold changes, p values and adjusted
p values.

With no additional arguments to results, the log2 fold change and Wald test p 
value will be for the last variable in the design formula, and if this is a 
factor, the comparison will be the last level of this variable over the 
reference level

```{r}
coefList <- resultsNames(dds) # lists the coefficients
coefList
```

### Calculate the 'un shrunk' log fold change
```{r}
# see vignette for how to specify coefficient or contrast. contrast is the factor
# level we want to compare against, it is the control contrast also effect the log
# fold change value if count is zero see ?results Note that the results function
# automatically performs independent filtering based on the mean of normalized
# counts for each gene, optimizing the number of genes which will have an
# adjusted p value below a given FDR cutoff, alpha By default the argument alpha
# is set to . If the adjusted p value cutoff will be a value other than , alpha
# should be set to that value:
# we want 0.05 in either tail. 
#
# we can use lfcThreshold to specify a minimum biologically meaningful effect size
res.unShrunk <- results(dds, name=coefList[2], lfcThreshold=1)
head(res.unShrunk)
```

```{r}
summary( res.unShrunk )
```
### Save LFC data
```{r}
saveLCF <- function(res, dataOutDir, dataSetName, nameStr, shrinkType="") {
  if (shrinkType == "") {
    metaName <- paste0(dataSetName, ".", nameStr, ".meta.txt")
    dataName <- paste0(dataSetName, ".", nameStr, ".csv")
  } else {
    metaName <- paste0(dataSetName, ".", shrinkType, ".", nameStr, ".meta.txt")
    dataName <- paste0(dataSetName, ".", shrinkType, ".", nameStr, ".csv")
  }
  
  # use capture.output to save output from show(res.lfcShrink)
  # https://www.gastonsanchez.com/visually-enforced/how-to/2014/12/21/capturing-R-printed-output/
  metaDataFile <- file.path(dataOutDir, metaName)
  capture.output(res, file=metaDataFile)
  cat(sprintf("LCF meta data saved to \n%s\n", metaDataFile))
  
  # save the actual LCF values in csv file
  dataFile <- file.path(dataOutDir, dataName)
  write.csv(res, dataFile)
  cat(sprintf("LCF data saved to \n%s\n", dataFile))
}
```

```{r}
saveLCF(res.unShrunk, dataOutDir, dataSetName, "unShrunk", shrinkType="")
```

### Calculate the 'shrunk' log fold change
```{r}
# If we just use results() and plotMA() many large LFC which are not significant
# these obtain a large LFC because of imprecise log counts
# 
# use lfcShrink() to shrink log fold changes association with condition:
# Shrinkage of effect size (LFC estimates) is useful for visualization and 
# ranking of genes.
# we can use lfcThreshold to specify a minimum biologically meaningful effect size

# type  argument "apeglm" is the adaptive Student's t prior shrinkage estimator
# from the 'apeglm' package; "ashr" is the adaptive shrinkage estimator from the
# 'ashr' package, using a fitted mixture of normals prior - see the Stephens
# (2016) reference below for citation; "normal" is the 2014 DESeq2 shrinkage
# estimator using a Normal prior;

# https://rdrr.io/bioc/DESeq2/man/lfcShrink.html
shrinkageType <- "apeglm"
res.lfcShrink <- lfcShrink(dds, coef=coefList[2], type=shrinkageType, lfcThreshold=1)
head(res.lfcShrink)
```
```{r}
summary(res.lfcShrink)
```

### Save LFC data
```{r}
saveLCF(res.lfcShrink, dataOutDir, dataSetName, "lfcShrink", shrinkType=shrinkageType)
```


## Explore results
```{r}
MAPlotHack <- function() {
# use par to create subplots https://www.datamentor.io/r-programming/subplot/
par(mfrow=c(1,2))
title <- paste(dataSetName, "LFC No shrinkage", sep="\n")
plotMA(res.unShrunk, ylim=c(-5,5), main=title)

title <- paste(dataSetName, paste("LFC", shrinkageType, "shrinkage"), sep="\n")
# plotMA(res.lfcShrink, ylim=c(-5,5), main="pancreas.plasma.ev.long.RNA\n LFG apeglm shrinkage", alpha=0.01)
plotMA(res.lfcShrink, ylim=c(-5,5), main=title, alpha=0.01)

}

# save plots http://www.sthda.com/english/wiki/creating-and-saving-graphs-r-base-graphs
# Open a PNG file
imgOutfile <- file.path(imgDir, paste0(dataSetName, "_MA.png"))
png(imgOutfile) 

MAPlotHack()

# close the plot file
dev.off() 
cat(sprintf("plotMA images saved to\n%s", imgOutfile))

# plot second time so that image is displayed in notebook
MAPlotHack()
```
```{r}
# save plots http://www.sthda.com/english/wiki/creating-and-saving-graphs-r-base-graphs
# Open a PNG file
imgOutfile <- file.path(imgDir, paste0(dataSetName, "_dispEsts.png"))
png(imgOutfile) 

plotDispEsts(dds)

# close the plot file
dev.off() 
cat(sprintf("plotMA images saved to\n%s", imgOutfile))

# plot second time so it is displayed in notebook
plotDispEsts(dds)
```


# clean up
```{r}
endTime <- Sys.time()
print( endTime - startTime )
```

```{r}
# AEDWIP Hack to get meta data

capOutFile <- "aedwip.mcols.txt"
cat("mcols(res.unShrunk)\n", file=capOutFile)
capture.output(mcols(res.unShrunk), file=capOutFile, append = TRUE)

cat("\n\nmcols(res.lfcShrink) shrinkageType=apeglm \n", 
               file=capOutFile, append = TRUE)
capture.output(mcols(res.lfcShrink), file=capOutFile, append = TRUE)
```

