# initial data view with Roman

Goal: simple data exploration of extra cellular RNA
ref: BME notebook page 89 5/5 1:1

- what biotypes are represented?
  * types are labels like "micro rna", "non coding rna"
  * see [https://www.gencodegenes.org/](https://www.gencodegenes.org/) for annotation

- what transposable elements are derived

```
/public/groups/kimlab/kras.ipsc/exo.data
```

csv file  output of differential seq "DESEQ 2" [https://bioconductor.org/packages/release/bioc/html/DESeq.html](https://bioconductor.org/packages/release/bioc/html/DESeq.html). 

```
day.5.de.seq.norm.counts.csv
```


alu.rna.edit.out/

- rna editing pipe
  * 'alu' "transpostable element, alu is the major target for edting in cells
  * if editied marked as self, else consider not self

bam.files/

(base) [aedavids@courtyard exo.data]$ zcat bam.files/ctrl.1.out.bam | head
BAM?@HD	VN:1.4
@SQ	SN:chr1	LN:248956422
@SQ	SN:chr2	LN:242193529
@SQ	SN:chr3	LN:198295559
@SQ	SN:chr4	LN:190214555
@SQ	SN:chr5	LN:181538259
@SQ	SN:chr6	LN:170805979
@SQ	SN:chr7	LN:159345973
@SQ	SN:chr8	LN:145138636
@SQ	SN:chr9	LN:138394717

/public/groups/kimlab/kras.ipsc/exo.data/kim_2019_09_26_gen1c_exo_rna ??? permission denied


issue with differential expersion: data. DESQ 2 does not find anything that is signifigantly different in our libraries

```
(base) [aedavids@courtyard kim_2019_09_26_gen1c_exo_rna]$ ls 
ctrl.1
gencode.salmon.out/           
output_forward_unpaired.fq.gz  
quantFiles/                            
scram_gen1c_exo1_S133_R2_001.fastq.gz
gencode.te.locus.salmon.out/  
output_reverse_paired.fq.gz    
rna.edit.star.out/                     
star.out/
output_forward_paired.fq.gz   
output_reverse_unpaired.fq.gz  
scram_gen1c_exo1_S133_R1_001.fastq.gz
```
(base) [aedavids@courtyard kim_2019_09_26_gen1c_exo_rna]$

2 salmon output file
  - transcriptom aligment, fast not aligned to Whole Genome
  - one with gencode annotation
  - other with gencode annotation + TE seq
  - they have alignment and abundace information
  
- star alignment
   - aedwp us start
   - classic genomic alignment
    - splice aware
    - bam file
    - output file has 2 output file
    - just alignment, not quantified
    - tool htseq ?
    
```
(base) [aedavids@courtyard star.out]$ pwd
/public/groups/kimlab/kras.ipsc/exo.data/kim_2019_09_26_gen1c_exo_rna/ctrl.1/star.out
(base) [aedavids@courtyard star.out]$ zcat ctrl.1.out.bam | head
BAM"@HD	VN:1.4
@SQ	SN:chr1	LN:248956422
@SQ	SN:chr2	LN:242193529
@SQ	SN:chr3	LN:198295559
@SQ	SN:chr4	LN:190214555
@SQ	SN:chr5	LN:181538259
@SQ	SN:chr6	LN:170805979
@SQ	SN:chr7	LN:159345973
@SQ	SN:chr8	LN:145138636
@SQ	SN:chr9	LN:138394717
```

do not parse bam, there are tools to quantify, salmon, ...
Roman created

```
(base) [aedavids@courtyard exo.data]$ pwd
/public/groups/kimlab/kras.ipsc/exo.data
(base) [aedavids@courtyard exo.data]$ ls -ld exo.te.locus.normalized.counts.csv 
-rw-r--r-- 1 rreggiar kimlab 3.6M Feb 17 15:07 exo.te.locus.normalized.counts.csv
```

each row is gene, see top for col names. each col is a library
so we have 6 librarys
each row is a specific gen, and its avlue in each library, normalized
interlibrary nromalization

```
base) [aedavids@courtyard exo.data]$ tail exo.te.locus.normalized.counts.csv
annotated abundance DESeq 2
"ZWINT",6.62369828553881,20.8966862763846,10.4323477491627,15.126700797512,8.0690002877522,4.15215144352486
"ZXDA",0,0,0,0.450580834346292,1.12352482897828,0
"ZXDB",0,0,0,0.485573406931516,0,0
"ZXDC",1.7306563064273,0.680458356936635,1.54389996008803,1.78965570176091,2.20095775952764,0.837796452065375
"ZYG11A",5.78073297946698,2.86777406754007,3.78871768579502,2.38094206521651,2.3372262157004,1.23598751038551
"ZYG11AP1",1.35717989250784,0,0,0,0,0
"ZYG11B",5.96793016613246,6.08333299587039,6.43590795408081,2.7163899108026,6.81998411106,10.1879773573553
"ZYX",4.23252562877027,4.53305281569661,11.3165414010699,6.50750743805843,2.091005101468,2.274549618252
"ZYXP1",0,0,0,0,0,0
"ZZEF1",9.47301597561893,10.5061022408547,3.89858891466308,3.66198826180922,13.7654874535881,9.37963721631963
(base) [aedavids@courtyard exo.data
```

aedwip
