{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0d4c8a8e-747f-42db-83e3-02dd43fbbf64",
   "metadata": {},
   "source": [
    "# Bar Charts\n",
    "Andrew E. Davidson  \n",
    "aedavids@ucsc.edu\n",
    "\n",
    "Copyright (c) 2020-2023, Regents of the University of California All rights reserved. https://polyformproject.org/licenses/noncommercial/1.0.0\n",
    "\n",
    "ref:\n",
    "- /extraCellularRNA/terra/jupyterNotebooks/cibersort/volcanoPlots.html\n",
    "volcano plots I made for Daniel Dec/2022 Stanford A.I. conference. The plots show differentialy express genes with TE gene plotted in red\n",
    "\n",
    "**bar chart over vie generated by chatGPT4** \n",
    "- x axis categorys Lung, LUAD, ...\n",
    "- y ratio = (number of repeats) / (number that are not repeats)\n",
    "- standard error of a proportion is ( p(1-p) / n) **0.5 ; AEDWIP this looks like the std\n",
    "  * n = number of observations\n",
    "  * p is proportion of the numerator of the ration. ie proportion of repeats\n",
    "- error bar range = standard error * Z\n",
    "  * Z = Z-score corresponding to your confidence level (e.g., 1.96 for 95% confidence)\n",
    " \n",
    "**proportion stats**  \n",
    "[6.3: The Sample Proportion](https://stats.libretexts.org/Bookshelves/Introductory_Statistics/Introductory_Statistics_(Shafer_and_Zhang)/06%3A_Sampling_Distributions/6.03%3A_The_Sample_Proportion#:~:text=at%20least%2030.-,The%20Sampling%20Distribution%20of%20the%20Sample%20Proportion,P%3D%E2%88%9Apqn.)\n",
    "\n",
    "p = num of positive observations / total number of observations  \n",
    "q =  1 - p  \n",
    "n = number of observations  \n",
    "std = sqrt( pq / n )\n",
    "\n",
    "**error bar choices**  \n",
    "1. standard deviation \n",
    "2. standard error (the standard deviation of the averages)\n",
    "3. confidence interval  \n",
    "    C.I = observed metric +/- the margin of error\n",
    "    + factors affecting CI  \n",
    "      a. number of observations  \n",
    "      b. variablity  \n",
    "      c. The confidence level: how much you can trust your confidence-interval calculation\n",
    "    + [Khan Acadamey Confidence intervals and margin of error](https://www.khanacademy.org/math/ap-statistics/xfb5d8e68:inference-categorical-proportions/introduction-confidence-intervals/v/confidence-intervals-and-margin-of-error) \n",
    "5. aaa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "696aca97-f590-4238-ab59-bcb1ff9517ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p : 0.4\n",
      "se :0.06928203230275509\n",
      "errorBar : [0.13579278331339997, -0.13579278331339997]\n",
      "y : 0.6666666666666667 20/50 : 0.6666666666666666\n"
     ]
    }
   ],
   "source": [
    "# example of blot data generated by chatGPT4\n",
    "n= 50\n",
    "p = 20 / n\n",
    "print(f'p : {p}')\n",
    "se = (p *(1-p)/n)**0.5\n",
    "print(f'se :{se}')\n",
    "\n",
    "# error bar Error Bars (95% CI)\n",
    "z = 1.96\n",
    "errorBars = [se*z, -se*z]\n",
    "print( f'errorBar : {errorBars}' )\n",
    "\n",
    "y = p / (1-p)\n",
    "print(f'y : {y} 20/50 : {20/30}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "f65d268c-6db3-4194-a4ff-a38751a0ec0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outDir:\n",
      "/private/home/aedavids/extraCellularRNA/terra/GTExTCGA_Release/jupyterNotebooks/barCharts.out\n"
     ]
    }
   ],
   "source": [
    "import ipynbname\n",
    "from IPython.display import display\n",
    "import pandas as pd\n",
    "# display all rows\n",
    "pd.set_option('display.max_rows', None)\n",
    "\n",
    "# https://joelmccune.com/pandas-dataframe-to-markdown/\n",
    "from pandas.io.clipboards import to_clipboard\n",
    "\n",
    "import pathlib as pl\n",
    "import pprint as pp\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "notebookName = ipynbname.name()\n",
    "notebookPath = ipynbname.path()\n",
    "notebookDir = os.path.dirname(notebookPath)\n",
    "\n",
    "outDir = f'{notebookDir}/{notebookName}.out'\n",
    "os.makedirs(outDir, exist_ok=True)\n",
    "print(f'outDir:\\n{outDir}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41fcc706-3110-4b33-8199-b10784e94db8",
   "metadata": {},
   "source": [
    "## How the te gene names work?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8fc505ac-d1d7-4ed4-998d-d2c00e7edede",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "/private/groups/kimlab/genomes.annotations/genomes.annotations/gencode.39/named_ucsc.rmsk.insert.info.txt\n",
      "the insert column is repeat name\n",
      "(15625, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>insert</th>\n",
       "      <th>clade</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(TAACCC)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TAR1</td>\n",
       "      <td>Satellite</td>\n",
       "      <td>telo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1MC5a</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MER5B</td>\n",
       "      <td>DNA</td>\n",
       "      <td>hAT-Charlie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR3</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      insert          clade         family\n",
       "0  (TAACCC)n  Simple_repeat  Simple_repeat\n",
       "1       TAR1      Satellite           telo\n",
       "2     L1MC5a           LINE             L1\n",
       "3      MER5B            DNA    hAT-Charlie\n",
       "4       MIR3           SINE            MIR"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "refRootDir = pl.Path('/private/groups/kimlab/genomes.annotations/genomes.annotations/gencode.39')\n",
    "teInfoFilePath = refRootDir.joinpath('named_ucsc.rmsk.insert.info.txt')\n",
    "teNameDF = pd.read_csv(teInfoFilePath, sep=\"\\t\")\n",
    "\n",
    "print( f\"\\n{teInfoFilePath}\" )\n",
    "print( \"the insert column is repeat name\")\n",
    "print(teNameDF.shape)\n",
    "display(teNameDF.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "68c5c037-ffa7-4d6e-abc2-2d8808a473eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Simple_repeat', 'Satellite', 'LINE', 'DNA', 'SINE', 'LTR',\n",
       "       'Low_complexity', 'LTR?', 'snRNA', 'tRNA', 'DNA?', 'Retroposon',\n",
       "       'srpRNA', 'rRNA', 'Unknown', 'RC', 'scRNA', 'RNA', 'RC?', 'SINE?'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniqueCladeNP = teNameDF.loc[:, 'clade'].unique()\n",
    "uniqueCladeNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "859dbb5d-fc1f-408b-bb42-145218ef2ac8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(67,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['Simple_repeat', 'telo', 'L1', 'hAT-Charlie', 'MIR', 'L2', 'CR1',\n",
       "       'ERVL-MaLR', 'Alu', 'hAT', 'ERVL', 'LTR', 'ERV1', 'hAT-Tip100',\n",
       "       'Low_complexity', 'TcMar-Tigger', 'ERVL?', 'ERV1?', 'RTE-X',\n",
       "       'LTR?', 'ERVK', 'snRNA', 'MULE-MuDR', 'tRNA', 'DNA?', 'Gypsy',\n",
       "       'hAT-Blackjack', 'SVA', 'Satellite', 'srpRNA', 'hAT-Ac', 'rRNA',\n",
       "       'TcMar-Tc2', 'tRNA-Deu', 'RTE-BovB', '5S-Deu-L2', 'PiggyBac',\n",
       "       'hAT?', 'Unknown', 'TcMar-Mariner', 'Helitron', 'Gypsy?',\n",
       "       'tRNA-RTE', 'hAT-Tip100?', 'DNA', 'Penelope', 'Dong-R4', 'scRNA',\n",
       "       'RNA', 'PiggyBac?', 'centr', 'TcMar?', 'Helitron?', 'TcMar',\n",
       "       'hAT-Tag1', 'Merlin', 'TcMar-Pogo', 'SINE?', 'PIF-Harbinger',\n",
       "       'acro', 'Kolobok', 'TcMar-Tc1', 'Crypton', 'I-Jockey',\n",
       "       'MULE-MuDR?', 'L1-Tx1', 'subtelo'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniqueFamilyNP = teNameDF.loc[:, 'family'].unique()\n",
    "print( uniqueFamilyNP.shape )\n",
    "uniqueFamilyNP "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3727d368-de83-4eeb-95c7-c04d577d9ed9",
   "metadata": {},
   "source": [
    "## data wrangling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f1ca7bc4-4187-41cb-91d9-1c56f4bc30c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_id,participant_id,category,gender,age,dataSet\n",
      "GTEX-1117F-0226-SM-5GZZ7,GTEX-1117F,Adipose_Subcutaneous,Female,66.0,GTEx\n",
      "GTEX-1117F-0526-SM-5EGHJ,GTEX-1117F,Artery_Tibial,Female,66.0,GTEx\n",
      "GTEX-1117F-0726-SM-5GIEN,GTEX-1117F,Heart_Atrial_Appendage,Female,66.0,GTEx\n",
      "GTEX-1117F-2826-SM-5GZXL,GTEX-1117F,Breast_Mammary_Tissue,Female,66.0,GTEx\n",
      "GTEX-1117F-3226-SM-5N9CT,GTEX-1117F,Brain_Cortex,Female,66.0,GTEx\n",
      "GTEX-111CU-0326-SM-5GZXO,GTEX-111CU,Lung,Male,57.0,GTEx\n",
      "GTEX-111CU-0426-SM-5GZY1,GTEX-111CU,Spleen,Male,57.0,GTEx\n",
      "GTEX-111CU-0526-SM-5EGHK,GTEX-111CU,Pancreas,Male,57.0,GTEx\n",
      "GTEX-111CU-0626-SM-5EGHL,GTEX-111CU,Esophagus_Muscularis,Male,57.0,GTEx\n"
     ]
    }
   ],
   "source": [
    "dataRoot = \"/private/groups/kimlab/GTEx_TCGA/groupbyGeneTrainingSets\"\n",
    "! head $dataRoot/GTEx_TCGA_TrainColData.csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "42d8f264-de60-4ff4-be60-ea8381364091",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "geneId,GTEX-1117F-0226-SM-5GZZ7,GTEX-1117F-0526-SM-5EGHJ\n",
      "(A)n,9,1\n",
      "(AAA)n,0,0\n",
      "(AAAAAAC)n,0,0\n",
      "(AAAAAAG)n,0,0\n",
      "(AAAAAAT)n,0,0\n",
      "(AAAAAC)n,0,0\n",
      "(AAAAACA)n,0,0\n",
      "(AAAAACC)n,0,0\n",
      "(AAAAACT)n,0,0\n"
     ]
    }
   ],
   "source": [
    "! head $dataRoot/GTEx_TCGA_TrainGroupby.csv | cut -d , -f 1,2,3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f8dd506-778c-4362-8f2b-58bcf08e4d0d",
   "metadata": {},
   "source": [
    "```\n",
    "how many columns in training data set\n",
    "head -n 1 $dataRoot/GTEx_TCGA_TrainGroupby.csv | comma2newLine | wc -l\n",
    "\n",
    "(extraCellularRNA) aedavids@mustard $ head -n 1 GTEx_TCGA_TrainGroupby.csv | comma2newLine | wc -l\n",
    "15802\n",
    "\n",
    "(extraCellularRNA) aedavids@mustard $ head -n 1 GTEx_TCGA_ValidateGroupby.csv | comma2newLine | wc -l\n",
    "5269\n",
    "\n",
    "(extraCellularRNA) aedavids@mustard $ head -n 1 GTEx_TCGA_TestGroupby.csv | comma2newLine | wc -l\n",
    "5269\n",
    "\n",
    "(extraCellularRNA) aedavids@mustard $ wc -l GTEx_TCGA_*Groupby.csv \n",
    "     74778 GTEx_TCGA_TestGroupby.csv\n",
    "     74778 GTEx_TCGA_TrainGroupby.csv\n",
    "     74778 GTEx_TCGA_ValidateGroupby.csv\n",
    "    224334 total\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4323c232-2ff3-401b-9a78-6203d5527b85",
   "metadata": {},
   "source": [
    "### Create some plot test data\n",
    "- select 2lung and 2 LUAD sample from training data set\n",
    "- select 6 rows that do not have any zeros. We will say the top are repeats the bottom 6 are not repeats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "cdd8a074-f52b-44c8-907c-ca3d87c2bf9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/private/groups/kimlab/GTEx_TCGA/groupbyGeneTrainingSets/GTEx_TCGA_TrainColData.csv\n",
      "(15801, 6)\n",
      "\n",
      "/private/groups/kimlab/GTEx_TCGA/groupbyGeneTrainingSets/GTEx_TCGA_TrainGroupby.csv\n",
      "(74777, 15802)\n",
      "CPU times: user 3min 35s, sys: 13.8 s, total: 3min 49s\n",
      "Wall time: 3min 49s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "trainColDataPath   = f'{dataRoot}/GTEx_TCGA_TrainColData.csv'\n",
    "print(trainColDataPath)\n",
    "trainColDataDF = pd.read_csv(trainColDataPath)\n",
    "print(trainColDataDF.shape)\n",
    "\n",
    "print()\n",
    "trainCountPath   = f'{dataRoot}/GTEx_TCGA_TrainGroupby.csv'\n",
    "print(trainCountPath)\n",
    "trainCountDF = pd.read_csv(trainCountPath)\n",
    "print(trainCountDF.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "b9f712db-a283-43bc-b4e3-d1b0ca06ffad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['sample_id', 'participant_id', 'category', 'gender', 'age', 'dataSet'], dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GTEX-111CU-0326-SM-5GZXO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GTEX-111VG-0726-SM-5GIDC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample_id\n",
       "5   GTEX-111CU-0326-SM-5GZXO\n",
       "29  GTEX-111VG-0726-SM-5GIDC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13310</th>\n",
       "      <td>LUAD-05-4244-TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13311</th>\n",
       "      <td>LUAD-05-4249-TP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample_id\n",
       "13310  LUAD-05-4244-TP\n",
       "13311  LUAD-05-4249-TP"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# find 2 samples\n",
    "print(trainColDataDF.columns)\n",
    "selectLungRows = trainColDataDF.loc[:, 'category'] == \"Lung\"\n",
    "lungSampleIdsDF = trainColDataDF.loc[selectLungRows, ['sample_id']].head(n=2)\n",
    "display(lungSampleIdsDF)\n",
    "\n",
    "print()\n",
    "selectLUADRows = trainColDataDF.loc[:, 'category'] == \"LUAD\"\n",
    "LUADSampleIdsDF = trainColDataDF.loc[selectLUADRows, ['sample_id']].head(n=2)\n",
    "display(LUADSampleIdsDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "b1b34b47-a7dd-4345-ba57-f630bd53baf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GTEX-111CU-0326-SM-5GZXO', 'GTEX-111VG-0726-SM-5GIDC', 'LUAD-05-4244-TP', 'LUAD-05-4249-TP']\n",
      "(74777, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>GTEX-111CU-0326-SM-5GZXO</th>\n",
       "      <th>GTEX-111VG-0726-SM-5GIDC</th>\n",
       "      <th>LUAD-05-4244-TP</th>\n",
       "      <th>LUAD-05-4249-TP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(A)n</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(AAA)n</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(AAAAAAC)n</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(AAAAAAG)n</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(AAAAAAT)n</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       geneId  GTEX-111CU-0326-SM-5GZXO  GTEX-111VG-0726-SM-5GIDC  \\\n",
       "0        (A)n                         2                        12   \n",
       "1      (AAA)n                         0                         0   \n",
       "2  (AAAAAAC)n                         0                         0   \n",
       "3  (AAAAAAG)n                         0                         0   \n",
       "4  (AAAAAAT)n                         0                         0   \n",
       "\n",
       "   LUAD-05-4244-TP  LUAD-05-4249-TP  \n",
       "0               11                2  \n",
       "1                0                0  \n",
       "2                0                0  \n",
       "3                0                0  \n",
       "4                0                0  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testSampleCols = lungSampleIdsDF['sample_id'].to_list() + LUADSampleIdsDF['sample_id'].to_list()\n",
    "print(testSampleCols)\n",
    "plotTestCountsDF = trainCountDF.loc[:, ['geneId'] +  testSampleCols]\n",
    "print(plotTestCountsDF.shape)\n",
    "plotTestCountsDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "7d8574ab-40f7-4e42-bf7f-f666c4f64984",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numSampleCols : 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>GTEX-111CU-0326-SM-5GZXO</th>\n",
       "      <th>GTEX-111VG-0726-SM-5GIDC</th>\n",
       "      <th>LUAD-05-4244-TP</th>\n",
       "      <th>LUAD-05-4249-TP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(A)n</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  geneId  GTEX-111CU-0326-SM-5GZXO  GTEX-111VG-0726-SM-5GIDC  LUAD-05-4244-TP  \\\n",
       "0   (A)n                         2                        12               11   \n",
       "\n",
       "   LUAD-05-4249-TP  \n",
       "0                2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def selectGenesWithCounts(df : pd.DataFrame) -> pd.DataFrame:\n",
    "    '''\n",
    "    make bar code plot testing easier. select rows that do not have zeros\n",
    "    '''\n",
    "    # create a logical data frame\n",
    "    nonZeroCountsDF = df != 0\n",
    "    print\n",
    "\n",
    "    byRow = 1\n",
    "    nonZeroCountsDF['rowSum'] = nonZeroCountsDF.sum(axis=byRow)\n",
    "\n",
    "    numSampleCols = df.shape[1]\n",
    "    print(f'numSampleCols : {numSampleCols}')\n",
    "    selectGenesWithOutZerosRow = nonZeroCountsDF['rowSum'] == numSampleCols\n",
    "\n",
    "    retDF = df.loc[selectGenesWithOutZerosRow, :]\n",
    "\n",
    "    return retDF\n",
    "\n",
    "def testSelectGenesWithCounts():\n",
    "    retDF = selectGenesWithCounts( plotTestCountsDF.head() )\n",
    "    display(retDF)\n",
    "\n",
    "testSelectGenesWithCounts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "c608d6f3-c659-4624-941a-ba0ecd73a501",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "numSampleCols : 5\n",
      "{'LUAD':         geneId  LUAD-05-4244-TP  LUAD-05-4249-TP  rowSum  isRepeat\n",
      "0         (A)n               11                2      13      True\n",
      "102    (AAAG)n                4                1       5      True\n",
      "528    (AAGG)n                5                1       6      True\n",
      "841    (AATG)n               31                3      34     False\n",
      "885   (AATGG)n              229               44     273     False\n",
      "1043     (AC)n              685              393    1078     False,\n",
      " 'Lung':         geneId  GTEX-111CU-0326-SM-5GZXO  GTEX-111VG-0726-SM-5GIDC  rowSum  \\\n",
      "0         (A)n                         2                        12      14   \n",
      "102    (AAAG)n                         1                        11      12   \n",
      "528    (AAGG)n                         3                         4       7   \n",
      "841    (AATG)n                         4                         8      12   \n",
      "885   (AATGG)n                        38                        22      60   \n",
      "1043     (AC)n                      1112                      1006    2118   \n",
      "\n",
      "      isRepeat  \n",
      "0         True  \n",
      "102       True  \n",
      "528       True  \n",
      "841      False  \n",
      "885      False  \n",
      "1043     False  }\n"
     ]
    }
   ],
   "source": [
    "def createBarChartPlotData(plotTestCountsDF) :\n",
    "    nonZeroCountsDF = selectGenesWithCounts( plotTestCountsDF )\n",
    "    nonZeroCountsDF = nonZeroCountsDF.head(n=6)\n",
    "  \n",
    "    typeDict = dict()\n",
    "    typeDict['Lung'] = nonZeroCountsDF.loc[:, ['geneId'] + lungSampleIdsDF['sample_id'].to_list()]\n",
    "    typeDict['LUAD'] = nonZeroCountsDF.loc[:, ['geneId'] + LUADSampleIdsDF['sample_id'].to_list()]\n",
    "    \n",
    "    #print( pp.pformat(typeDict) )\n",
    "    \n",
    "    # sum the transcript counts\n",
    "    byRow = 1\n",
    "    for key, df in typeDict.items():\n",
    "        # all colunms except geneId\n",
    "        sampleCols = ~ df.columns.isin( ['geneId'] )\n",
    "        df['rowSum'] = df.loc[:, sampleCols].sum(axis=byRow)\n",
    "    \n",
    "    # print('\\n******** sum rows')\n",
    "    # print( pp.pformat(typeDict) )\n",
    "    \n",
    "    # assume the first 3 cols are repeats\n",
    "    for key, df in typeDict.items():\n",
    "        # all colunms except geneId\n",
    "        sampleCols = ~ df.columns.isin( ['geneId'] )\n",
    "        n = int(df.shape[0] / 2)\n",
    "        isRepeat = [True]*n + [False]*n\n",
    "        df['isRepeat'] = isRepeat\n",
    "    \n",
    "    # print('\\n******** sum rows')\n",
    "    # print( pp.pformat(typeDict) )\n",
    "\n",
    "    return typeDict\n",
    "\n",
    "barChartDict = createBarChartPlotData(plotTestCountsDF)\n",
    "print( pp.pformat(barChartDict) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "7f3203fa-b8b5-475e-840d-0def3a845806",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LUAD': {'df':         geneId  LUAD-05-4244-TP  LUAD-05-4249-TP  rowSum  isRepeat\n",
      "0         (A)n               11                2      13      True\n",
      "102    (AAAG)n                4                1       5      True\n",
      "528    (AAGG)n                5                1       6      True\n",
      "841    (AATG)n               31                3      34     False\n",
      "885   (AATGG)n              229               44     273     False\n",
      "1043     (AC)n              685              393    1078     False,\n",
      "          'errorBars': [0.03586663090894229, -0.03586663090894229],\n",
      "          'numColNameCounts': 24,\n",
      "          'p': 0.017033356990773598,\n",
      "          'se': 0.01829930148415423,\n",
      "          'totalNumCounts': 1409,\n",
      "          'y': 0.017033356990773598,\n",
      "          'z': 1.96},\n",
      " 'Lung': {'df':         geneId  GTEX-111CU-0326-SM-5GZXO  GTEX-111VG-0726-SM-5GIDC  rowSum  \\\n",
      "0         (A)n                         2                        12      14   \n",
      "102    (AAAG)n                         1                        11      12   \n",
      "528    (AAGG)n                         3                         4       7   \n",
      "841    (AATG)n                         4                         8      12   \n",
      "885   (AATGG)n                        38                        22      60   \n",
      "1043     (AC)n                      1112                      1006    2118   \n",
      "\n",
      "      isRepeat  \n",
      "0         True  \n",
      "102       True  \n",
      "528       True  \n",
      "841      False  \n",
      "885      False  \n",
      "1043     False  ,\n",
      "          'errorBars': [0.02864506697601569, -0.02864506697601569],\n",
      "          'numColNameCounts': 33,\n",
      "          'p': 0.010796221322537112,\n",
      "          'se': 0.014614830089803924,\n",
      "          'totalNumCounts': 2223,\n",
      "          'y': 0.014844804318488529,\n",
      "          'z': 1.96}}\n"
     ]
    }
   ],
   "source": [
    "def calcBarChartXY(barChartDict : dict[key, pd.DataFrame], \n",
    "                   colName : str) -> dict :\n",
    "# def calcBarChartXY(barChartDict , \n",
    "#                    colName )  :\n",
    "    '''\n",
    "    arguments\n",
    "        barChartDict\n",
    "\n",
    "        colName : str \n",
    "            column to calculate the proportion on. Example 'isRepeat'\n",
    "\n",
    "    output \n",
    "        update input dictionary \n",
    "        df \n",
    "        y = proportion example number of repeats / total number\n",
    "        e = error bar @ (95% CI)\n",
    "        x = category\n",
    "        totalNumCounts\n",
    "        numColNameCounts\n",
    "    '''\n",
    "    retDict = dict()\n",
    "    for key,df in barChartDict.items():\n",
    "        plotDataDict = dict()\n",
    "        plotDataDict['df'] = df\n",
    "        sumDF = df.loc[:,[colName, 'rowSum']].groupby([colName]).sum()\n",
    "\n",
    "        totalNumCounts = sumDF['rowSum'].sum()\n",
    "        plotDataDict['totalNumCounts'] = totalNumCounts\n",
    "\n",
    "        # get the number of  counts for colName\n",
    "        selectRow = sumDF.index # the index is a boolean. There are only 2 rows\n",
    "        numColNameCounts = sumDF.loc[selectRepeatRow , 'rowSum' ].values[0]\n",
    "        plotDataDict['numColNameCounts'] = numColNameCounts\n",
    "    \n",
    "        y = numColNameCounts / totalNumCounts\n",
    "        plotDataDict['y'] = y\n",
    "\n",
    "        # p is proportion for\n",
    "        p = numRepeatCounts / totalNumCounts\n",
    "        # print(f'\\np : {p}')\n",
    "        plotDataDict['p'] = p\n",
    "\n",
    "        # y and p should be equal. we just calculated them differently to add readablity\n",
    "\n",
    "        # calculate the standard error for the proportion\n",
    "        se = se = (p *(1-p)/n)**0.5\n",
    "        plotDataDict['se'] = se\n",
    "        \n",
    "        # error bar Error Bars (95% CI)\n",
    "        z = 1.96\n",
    "        plotDataDict['z'] = z\n",
    "        errorBars = [se*z, -se*z]\n",
    "        #print( f'errorBar : {errorBars}' )\n",
    "        plotDataDict['errorBars'] = errorBars\n",
    "\n",
    "        retDict[key] = plotDataDict\n",
    "\n",
    "\n",
    "    return retDict\n",
    "\n",
    "\n",
    "finalDict = calcBarChartXY(barChartDict, colName=\"isRepeat\")\n",
    "print( pp.pformat(finalDict) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f7a2c5d-21e3-42a8-95e0-b6dd3c904344",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2bf090c-9808-407a-b32e-41fe349cbf12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "6e1031c6-df37-4b25-b95d-01299d77d582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>LUAD-05-4244-TP</th>\n",
       "      <th>LUAD-05-4249-TP</th>\n",
       "      <th>rowSum</th>\n",
       "      <th>isRepeat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(A)n</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>(AAAG)n</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>(AAGG)n</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>841</th>\n",
       "      <td>(AATG)n</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>(AATGG)n</td>\n",
       "      <td>229</td>\n",
       "      <td>44</td>\n",
       "      <td>273</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1043</th>\n",
       "      <td>(AC)n</td>\n",
       "      <td>685</td>\n",
       "      <td>393</td>\n",
       "      <td>1078</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        geneId  LUAD-05-4244-TP  LUAD-05-4249-TP  rowSum  isRepeat\n",
       "0         (A)n               11                2      13      True\n",
       "102    (AAAG)n                4                1       5      True\n",
       "528    (AAGG)n                5                1       6      True\n",
       "841    (AATG)n               31                3      34     False\n",
       "885   (AATGG)n              229               44     273     False\n",
       "1043     (AC)n              685              393    1078     False"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "number of transcripts by repeat type\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isRepeat</th>\n",
       "      <th>rowSum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>0</td>\n",
       "      <td>1385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       isRepeat  rowSum\n",
       "False         0    1385\n",
       "True          3      24"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "totalNumCounts : 1409\n",
      "numRepeatCounts : 24\n",
      "\n",
      "y : 0.017033356990773598 numRepeatCounts/(totalNumCounts - numRepeatCounts-p) :0.017328519855595668\n",
      "\n",
      "p : 0.017033356990773598\n",
      "\n",
      "se : 0.01829930148415423\n",
      "errorBar : [0.03586663090894229, -0.03586663090894229]\n"
     ]
    }
   ],
   "source": [
    "lungAEDWIPDF = barChartDict['LUAD']\n",
    "display(lungAEDWIPDF)\n",
    "\n",
    "def \n",
    "lungSumDF = lungAEDWIPDF.loc[:,['isRepeat', 'rowSum']].groupby([isRepeat]).sum()\n",
    "\n",
    "print('\\nnumber of transcripts by repeat type')\n",
    "display(lungSumDF)\n",
    "\n",
    "totalNumCounts = lungSumDF['rowSum'].sum()\n",
    "print(f'\\ntotalNumCounts : {totalNumCounts}')\n",
    "\n",
    "# get the number of repeat counts\n",
    "selectRepeatRow = lungSumDF.index # the index is a boolean. There are only 2 rows\n",
    "numRepeatCounts = lungSumDF.loc[selectRepeatRow , 'rowSum' ].values[0]\n",
    "print(f'numRepeatCounts : {numRepeatCounts}')\n",
    "\n",
    "y = numRepeatCounts / totalNumCounts\n",
    "print(f'\\ny : {y} numRepeatCounts/(totalNumCounts - numRepeatCounts) :{numRepeatCounts/(totalNumCounts - numRepeatCounts)}')\n",
    "\n",
    "# p is proportion for\n",
    "p = numRepeatCounts / totalNumCounts\n",
    "print(f'\\np : {p}')\n",
    "\n",
    "\n",
    "# calculate the standard error for the proportion\n",
    "se = se = (p *(1-p)/n)**0.5\n",
    "print(f'\\nse : {se}')\n",
    "\n",
    "# error bar Error Bars (95% CI)\n",
    "z = 1.96\n",
    "errorBars = [se*z, -se*z]\n",
    "print( f'errorBar : {errorBars}' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5381772b-e4e7-4f26-97a5-d241a4478823",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "9c7fa6af-8daf-49eb-b895-1f1792b55a55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "ad5c7d3c-2c22-4eca-afe4-d9cdb229e152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([False, True], dtype='bool')"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b09ebe8b-f013-4f6c-8eda-7da398550157",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
