{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f07596e3",
   "metadata": {},
   "source": [
    "# Note using DESeq in a consistant way\n",
    "All of our DESeq work was not using R Notebooks. See extraCellularRNA/R/notebooks/ and extraCellularRNA/R/notebooks/reverseEngineerDESeq/ . We can not use DESeq and our current workflow to process the GTEx and TCGA data. These data sets are to big\n",
    "\n",
    "Part of our work around is run DESeq on a count matrix instead of using tximport to read roughly 29K quant.sf files.\n",
    "\n",
    "## Reverse engineer how DESeq works with salmon quant files\n",
    "additional references: extraCellularRNA/juypterNotebooks/spark\n",
    "\n",
    "extraCellularRNA/R/notebooks/kras.ipsc.DESeq.normalize.v2.nb.html\n",
    "\n",
    "```\n",
    "'KRAS.IPSC.day.5.gene'=list(\n",
    "    'rootDir' = krasDataRoot,\n",
    "    'findFilesFunc' = findDay5KrasIpscDataFilesGencode_v35_ucsc_rmsk_salmon_v1_3.0, \n",
    "    # colData is the meta data data frame\n",
    "    'getColDFFunc' = getKrasIPSCDay5ColDataDF,        \n",
    "    'tx2MappingFile'='gencode.v35.ucsc.rmsk.tx.to.gene.csv',\n",
    "    'tx2MappingDir' = file.path('/home/kimlab/genomes.annotations/gencode.35'),\n",
    "    'mapToIdName'='gene',\n",
    "    'normalizedFileName'='kras.ipsc.day.5.DESeq.normalize.v2',\n",
    "    'outputRoot' = \"/home/rstudio/extraCellularRNA/data/R/output\",\n",
    "    #'outputRoot' = \"/home/aedavids/extraCellularRNA/data/R/output\",\n",
    "    #'outputRoot' = \"/home/kimlab/aedavids/extraCellularRNA/data/R/output\",\n",
    "    'transcriptMapperDFFunc'= getTEGeneMapper,\n",
    "    # ignoreAfterBar logical, whether to split the tx id on the '|' character to \n",
    "    # facilitate matching with the tx id in tx2gene (default FALSE)\n",
    "    #'ignoreAfterBar'=FALSE #bug we do not get gene names in output\n",
    "    'ignoreAfterBar'=FALSE,\n",
    "    'txOut'=FALSE,\n",
    "    \n",
    "    # getColDFFunc returns a data frame with columns\n",
    "    #   factor sampleType ['bulk', 'exo']\n",
    "    #   factor treatment  ['ctrl', 'kras']\n",
    "    #   factor day        [5]\n",
    "    'designStr'=\"~ sampleType + treatment\"\n",
    "    ),\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fc0aa4c",
   "metadata": {},
   "source": [
    "# <span style=\"color:red\">main runNomalization()</span>\n",
    "\n",
    "```\n",
    "runNomalization <- function(parametersList, dataSetNamesList) {\n",
    "    #parameters <- parametersList$TE.gene\n",
    "    for (dataSetName in dataSetNamesList) {\n",
    "      parameters <- parametersList[[dataSetName]]\n",
    "      error = missingParmeters(dataSetName, parameters )\n",
    "      \n",
    "      if (error ) {\n",
    "        next # ie continue in most languages\n",
    "      }\n",
    "\n",
    "      retList    <- loadAndNormlize(parameters)\n",
    "      dsCountsDF <- retList$ddsCountsDF\n",
    "      dds        <- retList$dds\n",
    "      fileList   <- retList$fileList\n",
    "      \n",
    "      dsCountsDF <- fixColNames(fileList, dsCountsDF)\n",
    "      saveNormalizedDF(dsCountsDF, parameters)\n",
    "      \n",
    "      retList    <- runDifferentialExpression(dds, parameters)\n",
    "      coefList <- retList[ \"coefList\" ]\n",
    "      unShrunkResults <- retList[ \"unShrunkResults\" ]\n",
    "      lfcShrinkResults <- retList[ \"lfcShrinkResults\" ]\n",
    "\n",
    "      # save results\n",
    "      # warning(paste(\"\\nAEDWIP runNomalization coefList:\", coefList))\n",
    "      # warning(paste(\"\\nAEDWIP runNomalization coefList[1]:\", coefList[1]))\n",
    "      # cat(paste(\"\\n ???? [[1]]\", coefList[[1]]))\n",
    "      # # cat(paste(\"\\n ???? wtf\", coefList[2]))\n",
    "      # cat(paste(\"\\n ???? [[2]]\", coefList[[2]]))\n",
    "      # cat(paste(\"\\n ???? 1 1\", coefList[1][1]))\n",
    "      # cat(paste(\"\\n ???? 1 2\", coefList[1][2]))\n",
    "      # warning(paste(\"\\nAEDWIP runNomalization coefList[2]:\", coefList[2]))\n",
    "      \n",
    "      # coefList$coefList12]\n",
    "      # [1] \"Intercept\"\n",
    "      # coefList$coefList[2]\n",
    "      # [1] \"diseaseState_PDAC_vs_healthy\"\n",
    "      \n",
    "      nameAEDWIP = paste0(coefList$coefList[2], \".unShrunk\")\n",
    "      saveResults( unShrunkResults, nameAEDWIP, parameters )\n",
    "      nameAEDWIP = paste0(coefList$coefList[2], \".lfcShrink\")\n",
    "      saveResults( lfcShrinkResults, nameAEDWIP, parameters )\n",
    "      \n",
    "  }\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18f2ff57",
   "metadata": {},
   "source": [
    "## How does our mapping file deal with T.E.?\n",
    "\n",
    "```\n",
    "tx2MappingFile <- '/private/groups/kimlab/genomes.annotations/gencode.35/gencode.v35.ucsc.rmsk.tx.to.gene.csv'\n",
    "```\n",
    "\n",
    "<b>It is hard to see that this is a 2 column csv file, txtId, geneId</b>\n",
    "```\n",
    "(base) [aedavids@mustard gencode.35]$ head gencode.v35.ucsc.rmsk.tx.to.gene.csv \n",
    "ENST00000456328.2|ENSG00000223972.5|OTTHUMG00000000961.2|OTTHUMT00000362751.1|DDX11L1-202|DDX11L1|1657|processed_transcript|,DDX11L1\n",
    "ENST00000450305.2|ENSG00000223972.5|OTTHUMG00000000961.2|OTTHUMT00000002844.2|DDX11L1-201|DDX11L1|632|transcribed_unprocessed_pseudogene|,DDX11L1\n",
    "```\n",
    "\n",
    "\n",
    "```\n",
    "tail gencode.v35.ucsc.rmsk.tx.to.gene.csv\n",
    "hg38_rmsk_L1PA3_range=chr22_KI270739v1_random:49092-50070_5'pad=0_3'pad=0_strand=+_repeatMasking=none,L1PA3\n",
    "hg38_rmsk_ALR/Alpha_range=chr22_KI270739v1_random:50071-73985_5'pad=0_3'pad=0_strand=-_repeatMasking=none,ALR/Alpha\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c73a565d",
   "metadata": {},
   "source": [
    "# <span style=\"color:red\"> runNomalization -> loadAndNormlize </span>\n",
    "DESeqUtils.R\n",
    "\n",
    "0. <span style=\"color:red\"> state of our dds object, (DESeqDataSet), after running is </span>\n",
    "1. count matrix is set\n",
    "2. counts are grouped by gene id and counted\n",
    "3. counts have been converted to integers\n",
    "4. sanity check error if count < 0\n",
    "5. prefilter counts\n",
    "6. estimateSizeFactors have been calculated \n",
    "\n",
    "```\n",
    "loadAndNormlize <- function(parameters) {\n",
    "\n",
    "    ...\n",
    "    \n",
    " ddsList <- getDESeqDataSet(rootDir=parameters$rootDir,\n",
    "                             tx2MappingFilePath,\n",
    "                             findFilesFunc= parameters$findFilesFunc, \n",
    "                             getColDFFunc=parameters$getColDFFunc,\n",
    "                             design=design,\n",
    "                             transcriptMapperDF,\n",
    "                             ignoreAfterBar=ignoreAfterBar,\n",
    "                             txOut=parameters$txOut)\n",
    "  \n",
    "    line 313\n",
    "    prefilter\n",
    "    \n",
    "    \n",
    "    ddsCountsDF <- createComparableCounts( dds, filterRowCount=aedwip, filterSampleCount )\n",
    "  ddsCountsDF <- createComparableCounts( dds, filterRowCount=aedwip, filterSampleCount )\n",
    "  \n",
    "  \n",
    "  ret = list(ddsCountsDF,  dds, fileList) \n",
    "  names(ret) <- c(\"ddsCountsDF\", \"dds\", \"fileList\")\n",
    "  \n",
    "  return(ret)\n",
    " }\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23a342df",
   "metadata": {},
   "source": [
    "## runNomalization -> loadAndNormlize -> getDESeqDataSet\n",
    "```\n",
    "getDESeqDataSet <- function(rootDir,\n",
    "                            indexFilePath,\n",
    "                            findFilesFunc,\n",
    "                            getColDFFunc,\n",
    "                            design,\n",
    "                            transcriptMapperDF,\n",
    "                            ignoreAfterBar=TRUE,\n",
    "                            txOut=FALSE) {\n",
    "    \n",
    "  # tximport returns a list with matrices, \"abundance\", \"counts\", and \"length\", \n",
    "  # where the transcript level information is summarized to the gene-level. \n",
    "  # Typically, abundance is provided by the quantification tools as \n",
    "  # TPM (transcripts-per-million), while the counts are estimated counts \n",
    "  # (possibly fractional), and the \"length\" matrix contains the effective gene lengths\n",
    "    \n",
    "    # abundance == double, TPM column in salmon quant.sf\n",
    "    # counts    == double, numReads Column in salmon quant.sf\n",
    "    # length    == intetger , effective length column in salmon quant.sf\n",
    "    txi <- tximport(allSalmonQuantFiles, \n",
    "                  type=\"salmon\", \n",
    "                  tx2gene=transcriptMapperDF, \n",
    "                  ignoreAfterBar=ignoreAfterBar,\n",
    "                  txOut=txOut)\n",
    "    \n",
    "    # get the meta data for each column of count data.\n",
    "    colDF <- getColDFFunc(fileList)\n",
    "    \n",
    "    dds <- DESeqDataSetFromTximport(txi, colData=colDF, design)\n",
    "  \n",
    "  retList = List(dds, fileList)\n",
    "  names(retList) <- c(\"dds\", \"fileList\")\n",
    "  \n",
    "  return(retList) \n",
    "}\n",
    "    \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1538ec8",
   "metadata": {},
   "source": [
    "## runNomalization -> loadAndNormalize -> getDESeqDataSet -> tximport\n",
    "https://github.com/mikelove/tximport/blob/master/R/tximport.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaab23b5",
   "metadata": {},
   "source": [
    "```\n",
    "line 361: salmon processing\n",
    "txIdCol <- name\n",
    "abundanceCol <- Tpm\n",
    "countCol <- numReads\n",
    "lengthCol <- effectiveLength\n",
    "\n",
    "package readr is present\n",
    "line 368\n",
    "col.types <- readr::cols(\n",
    "        readr::col_character(),readr::col_integer(),readr::col_double(),readr::col_double(),readr::col_double()\n",
    "\n",
    "line 479, loop over all files\n",
    "  for (i in seq_along(files)) {\n",
    "    ...\n",
    "    \n",
    "      abundanceMatTx[,i] <- raw[[abundanceCol]]\n",
    "      countsMatTx[,i] <- raw[[countsCol]]\n",
    "      lengthMatTx[,i] <- raw[[lengthCol]]\n",
    "\n",
    "      ...\n",
    "line 547\n",
    "    }\n",
    "  ### --- END --- loop over files ###\n",
    "  \n",
    "  line 583\n",
    "      txi <- list(abundance=abundanceMatTx,\n",
    "                counts=countsMatTx,\n",
    "                length=lengthMatTx,\n",
    "                countsFromAbundance=countsFromAbundance)\n",
    "\n",
    "line 643\n",
    "txi[[\"countsFromAbundance\"]] <- NULL\n",
    "  txiGene <- summarizeToGene(txi, tx2gene, varReduce, ignoreTxVersion, ignoreAfterBar, countsFromAbundance)\n",
    "  return(txiGene)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5913c85b",
   "metadata": {},
   "source": [
    "##  runNomalization -> loadAndNormalize ->  getDESeqDataSet -> tximport -> summarizeToGene\n",
    "https://github.com/mikelove/tximport/blob/master/R/summarizeToGene.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4109a262",
   "metadata": {},
   "source": [
    "```\n",
    "object argument is the txi list create on line 583 in tximport\n",
    "line 32\n",
    "txId <- rownames(abundanceMatTx)\n",
    "stopifnot(all(txId == rownames(countsMatTx)))\n",
    "stopifnot(all(txId == rownames(lengthMatTx)))\n",
    "  \n",
    "line 47\n",
    "cleanTx2Gene removes duplicates and converts columns to factors\n",
    "tx2gene <- cleanTx2Gene(tx2gene)    \n",
    "\n",
    "\n",
    "lot of code to remove transcripts and genes not in the rownames of the matrices\n",
    "then create a geneId vector which aligns to the matrices\n",
    "```\n",
    "\n",
    "### group by gene and count\n",
    "\n",
    "```\n",
    "runsum(x, group, reoder=TRUE)\n",
    "\n",
    "group\t\n",
    "a vector or factor giving the grouping, with one element per row of x. Missing values will be treated as another group and a warning will be given.\n",
    "\n",
    "reorder\t\n",
    "if TRUE, then the result will be in order of sort(unique(group)), if FALSE, it will be in the order that groups were encountered.\n",
    "\n",
    "  countsMat <- rowsum(countsMatTx, geneId)\n",
    "  \n",
    "line  135\n",
    "# countsFromAbundance = 'no'\n",
    "out <- list(abundance=abundanceMat,\n",
    "                counts=countsMat,\n",
    "                length=lengthMat,\n",
    "                countsFromAbundance=countsFromAbundance)\n",
    "                \n",
    "line 141\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c18787da",
   "metadata": {},
   "source": [
    "## state after tximport\n",
    "- we have a list of 3 matrices. -\n",
    "- They have been \"grouped by gene\" and counted\n",
    "- <span style=\"color:red\"> counts have not been converted to integer</span>\n",
    "\n",
    "## runNomalization -> loadAndNormalize -> getDESeqDataSet -> DESeqDataSetFromTximport\n",
    "DESeqUtils.R\n",
    "-https://github.com/mikelove/DESeq2/blob/master/R/AllClasses.R\n",
    "<span style=\"color:red\"> # convert counts to integers </span>\n",
    "\n",
    "```\n",
    "line 408\n",
    "\n",
    "  # convert counts to integers\n",
    "  counts <- round(txi$counts)\n",
    "  \n",
    "  # help 'mode {base}'\n",
    "  # Get or set the type or storage mode of an object.\n",
    "  # this is not the 'mode' measure of central tendancy\n",
    "  # R does not have a standard in-built function to calculate mode. \n",
    "  mode(counts) <- \"integer\"\n",
    "\n",
    "  object <- DESeqDataSetFromMatrix(countData=counts, colData=colData, design=design, ...)\n",
    "\n",
    "```\n",
    "\n",
    "## runNomalization -> loadAndNormalize -> getDESeqDataSet -> DESeqDataSetFromTximport -> DESeqDataSetFromMatrix\n",
    "https://github.com/mikelove/DESeq2/blob/master/R/AllClasses.R\n",
    "\n",
    "<span style=\"color:red\">our DESeq data set only as count data </span>\n",
    "\n",
    "\n",
    "```\n",
    "line 334\n",
    "\n",
    "# santity check\n",
    "# check that these agree in number\n",
    "  stopifnot(ncol(countData) == nrow(colData))\n",
    "  \n",
    "  # we expect a matrix of counts, which are non-negative integers\n",
    "  countData <- as.matrix( countData )\n",
    "  \n",
    "  # check if the rownames of colData are simply in different order\n",
    "  # than the colnames of the countData, if so throw an error\n",
    "  # as the user probably should investigate what's wrong\n",
    "  \n",
    "  \n",
    "    se <- SummarizedExperiment(assays = SimpleList(counts=countData), colData = colData, ...)\n",
    "    \n",
    "    # DESeqDataSet does not do any numeric processing. Design is simply stored in returned object\n",
    "    # it enforces non-negative count values\n",
    "  object <- DESeqDataSet(se, design = design, ignoreRank)\n",
    "\n",
    "  return(object)\n",
    "  \n",
    "  \n",
    " DESeqDataSet()\n",
    "more sanity check\n",
    "# before validity check, try to convert assay to integer mode\n",
    "  if (any(is.na(assay(se))))\n",
    "    stop(\"NA values are not allowed in the count matrix\")\n",
    "  if (any(assay(se) < 0)) {\n",
    "    stop(\"some values in assay are negative\")\n",
    "  }\n",
    "  if (!is.integer(assay(se))) {\n",
    "    if (!is.numeric(assay(se))) {\n",
    "      stop(paste(\"counts matrix should be numeric, currently it has mode:\", mode(assay(se))))\n",
    "    }\n",
    "    if (any(round(assay(se)) != assay(se))) {\n",
    "      stop(\"some values in assay are not integers\")\n",
    "    }\n",
    "  \n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "## runNomalization -> loadAndNormalize  -> createComparableCounts\n",
    "\n",
    "DESeqUtils.R\n",
    "\n",
    "```\n",
    "line 11\n",
    "\n",
    "# prefilter counts \n",
    "# filter rows:\n",
    "# remove rows if they do not have a count of 5 or more in\n",
    "# 4 or more samples\n",
    "# dds is a DESeqDataSet object\n",
    "keep <- rowSums(counts(dds) >= filterRowCount) >= filterSampleCount\n",
    "table(keep)\n",
    "dds <- dds[keep,]  \n",
    "\n",
    "\n",
    "# https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/estimateSizeFactors\n",
    "  # This function estimates the size factors using the \"median ratio method\"\n",
    "  # described by Equation 5 in Anders and Huber (2010). The estimated size\n",
    "  # factors can be accessed using the accessor function sizeFactors\n",
    "  dds <- estimateSizeFactors( dds )\n",
    "  \n",
    "  aedwip\n",
    "```\n",
    "\n",
    "## runNomalization -> loadAndNormalize  -> createComparableCounts -> estimateSizeFactors\n",
    "\n",
    "line 367 https://github.com/mikelove/DESeq2/blob/master/R/methods.R\n",
    "\n",
    "```\n",
    "\n",
    "estimateSizeFactors.DESeqDataSet <- function(object, # object arg is a DESeqDataSet \n",
    "                                            type=c(\"ratio\",\"poscounts\",\"iterate\"),\n",
    "                                             locfunc=stats::median,\n",
    "                                             geoMeans, \n",
    "                                             controlGenes, \n",
    "                                             normMatrix, \n",
    "                                             quiet=FALSE) {\n",
    "\n",
    "\n",
    "\n",
    "line 397\n",
    "else if (missing(normMatrix)) {\n",
    "      sizeFactors(object) <- estimateSizeFactorsForMatrix(counts(object), locfunc=locfunc,\n",
    "                                                          geoMeans=geoMeans,\n",
    "                                                          controlGenes=controlGenes)\n",
    "    } else {\n",
    "      normalizationFactors(object) <- estimateNormFactors(counts(object),\n",
    "                                                          normMatrix=normMatrix,\n",
    "                                                          locfunc=locfunc,\n",
    "                                                          geoMeans=geoMeans,\n",
    "                                                          controlGenes=controlGenes)\n",
    "                                                          \n",
    "\n",
    "```\n",
    "\n",
    "<span style=\"color:red\">it is hard to find code, used print(funName) in RStudio console. It is very hard to follow code. I think this is the jest </span>\n",
    "\n",
    "```\n",
    "> print( estimateSizeFactorsForMatrix)\n",
    "function (counts, locfunc = stats::median, geoMeans, controlGenes, \n",
    "    type = c(\"ratio\", \"poscounts\")) \n",
    "{\n",
    "\n",
    "    if (type == \"ratio\") {\n",
    "            loggeomeans <- rowMeans(log(counts))\n",
    "        }\n",
    "        \n",
    "   sf <- if (missing(controlGenes)) {\n",
    "        apply(counts, 2, function(cnts) {\n",
    "            exp(locfunc((log(cnts) - loggeomeans)[is.finite(loggeomeans) & \n",
    "                cnts > 0]))\n",
    "        })\n",
    "    }\n",
    "    \n",
    "    \n",
    "  return sf\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "672c65ed",
   "metadata": {},
   "source": [
    "# <span style=\"color:red\"> runNomalization -> runDifferentialExpression(dds, parameters) </span>\n",
    "\n",
    "```\n",
    "  ddsDESeq <- DESeq(dds)\n",
    "\n",
    "```\n",
    "\n",
    "# <span style=\"color:red\"> runNomalization -> runDifferentialExpression(dds, parameters) -> DESeq() </span>\n",
    "core.R\n",
    "\n",
    "useful test function makeExampleDESeqDataSet\n",
    "    \n",
    "```\n",
    "# general outline of the internal function calls.\n",
    "# note: not all of these functions are exported.\n",
    "#\n",
    "# DESeq\n",
    "# |- estimateSizeFactors\n",
    "#    |- estimateSizeFactorsForMatrix\n",
    "# |- estimateDispersions\n",
    "#    |- estimateDispersionsGeneEst\n",
    "#       |- fitNbinomGLMs\n",
    "#          |- fitBeta (C++)\n",
    "#       |- fitDisp (C++)\n",
    "#    |- estimateDispersionsFit\n",
    "#    |- estimateDispersionsMAP\n",
    "#       |- estimateDispersionPriorVar\n",
    "#       |- fitDisp (C++)\n",
    "# |- nbinomWaldTest\n",
    "#    |- fitGLMsWithPrior\n",
    "#       |- fitNbinomGLMs\n",
    "#          |- fitBeta (C++)\n",
    "#       |- estimateBetaPriorVar\n",
    "#       |- fitNbinomGLMs\n",
    "#          |- fitBeta (C++)\n",
    "#\n",
    "```\n",
    "\n",
    "line 279:\n",
    "```\n",
    "DESeq <- function(object, # object is a DESeqDataSet\n",
    "                   test=c(\"Wald\",\"LRT\"),\n",
    "                  fitType=c(\"parametric\",\"local\",\"mean\", \"glmGamPoi\"),\n",
    "                  sfType=c(\"ratio\",\"poscounts\",\"iterate\"),\n",
    "                  betaPrior,\n",
    "                  full=design(object), \n",
    "                  reduced, \n",
    "                  quiet=FALSE,\n",
    "                  minReplicatesForReplace=7, \n",
    "                  modelMatrixType,\n",
    "                  useT=FALSE, \n",
    "                  minmu=if (fitType==\"glmGamPoi\") 1e-6 else 0.5,\n",
    "                  parallel=FALSE, \n",
    "                  BPPARAM=bpparam()) {\n",
    "                  \n",
    "    arguments values\n",
    "    test <- \"Wald\"\n",
    "    fitType <- parametric\n",
    "    stype <- ratio\n",
    "    betaPrior ???\n",
    "    full <- design\n",
    "    reduce ????\n",
    "    minReplicatesForReplace <- 7\n",
    "    modelMatixType ???\n",
    "    useT <- False\n",
    "    minmu <- 0.5\n",
    "    \n",
    "    code begins\n",
    "    \n",
    "    dispersionEstimator <- \"DESeq2\"\n",
    "    betaPrior <- FALSE\n",
    "    \n",
    "     # get rid of any NA in the mcols(mcols(object))\n",
    "  object <- sanitizeRowRanges(object)\n",
    "  \n",
    "  line 346\n",
    "  \n",
    "  ilne 376\n",
    "  \n",
    "  line 392\n",
    "      object <- estimateDispersions(object, fitType=fitType, quiet=quiet, modelMatrix=modelMatrix, minmu=minmu)\n",
    "\n",
    "    \n",
    "}\n",
    "\n",
    "sanitizeRowRanges <- function(object) {\n",
    "  if (is.null(mcols(mcols(object)))) {\n",
    "    mcols(mcols(object)) <- DataFrame(type=rep(\"input\",ncol(mcols(object))),\n",
    "                                      description=character(ncol(mcols(object))))\n",
    "  }\n",
    "  class(mcols(mcols(object))$type) <- \"character\"\n",
    "  class(mcols(mcols(object))$description) <- \"character\"\n",
    "  mcols(mcols(object))$type[ is.na(mcols(mcols(object))$type) ] <- \"\"\n",
    "  mcols(mcols(object))$description[ is.na(mcols(mcols(object))$description) ] <- \"\"\n",
    "  object\n",
    "}\n",
    " \n",
    "\n",
    "methods.R line 504\n",
    "estimateDispersions.DESeqDataSet <- function(object, fitType=c(\"parametric\",\"local\",\"mean\", \"glmGamPoi\"),\n",
    "                                             maxit=100, useCR=TRUE,\n",
    "                                             weightThreshold=1e-2,\n",
    "                                             quiet=FALSE, modelMatrix=NULL,\n",
    "                                             minmu=if (fitType==\"glmGamPoi\") 1e-6 else 0.5) {\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437df3f3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
